# ಡೇಟಾ ತಯಾರಿ

[ಮೂಲ ನೋಟ್ಬುಕ್ ಮೂಲ *ಡೇಟಾ ಸೈನ್ಸ್: ಪೈಥಾನ್ ಮತ್ತು ಮೆಷಿನ್ ಲರ್ನಿಂಗ್ ಸ್ಟುಡಿಯೋಗೆ ಡೇಟಾ ಸೈನ್ಸ್‌ಗೆ ಮೆಷಿನ್ ಲರ್ನಿಂಗ್ ಪರಿಚಯ* ಲೀ ಸ್ಟಾಟ್ ಅವರಿಂದ](https://github.com/leestott/intro-Datascience/blob/master/Course%20Materials/4-Cleaning_and_Manipulating-Reference.ipynb)

## `DataFrame` ಮಾಹಿತಿಯನ್ನು ಅನ್ವೇಷಿಸುವುದು

> **ಕಲಿಕೆಯ ಗುರಿ:** ಈ ಉಪವಿಭಾಗದ ಕೊನೆಯಲ್ಲಿ, pandas DataFrames ನಲ್ಲಿ ಸಂಗ್ರಹಿಸಲಾದ ಡೇಟಾ ಬಗ್ಗೆ ಸಾಮಾನ್ಯ ಮಾಹಿತಿಯನ್ನು ಕಂಡುಹಿಡಿಯಲು ನೀವು ಆರಾಮದಾಯಕವಾಗಿರಬೇಕು.

ನೀವು ನಿಮ್ಮ ಡೇಟಾವನ್ನು pandas ಗೆ ಲೋಡ್ ಮಾಡಿದ ನಂತರ, ಅದು ಬಹುಶಃ `DataFrame` ಆಗಿರುತ್ತದೆ. ಆದಾಗ್ಯೂ, ನಿಮ್ಮ `DataFrame` ನಲ್ಲಿ 60,000 ಸಾಲುಗಳು ಮತ್ತು 400 ಕಾಲಮ್‌ಗಳು ಇದ್ದರೆ, ನೀವು ಯಾವ ರೀತಿಯಲ್ಲಿ ನೀವು ಕೆಲಸ ಮಾಡುತ್ತಿರುವುದನ್ನು ಅರ್ಥಮಾಡಿಕೊಳ್ಳಲು ಪ್ರಾರಂಭಿಸುತ್ತೀರಿ? ಭಾಗ್ಯವಶಾತ್, pandas ಕೆಲವು ಅನುಕೂಲಕರ ಸಾಧನಗಳನ್ನು ಒದಗಿಸುತ್ತದೆ, ಇದು `DataFrame` ಬಗ್ಗೆ ಒಟ್ಟು ಮಾಹಿತಿಯನ್ನು ಮತ್ತು ಮೊದಲ ಕೆಲವು ಮತ್ತು ಕೊನೆಯ ಕೆಲವು ಸಾಲುಗಳನ್ನು ತ್ವರಿತವಾಗಿ ನೋಡಲು ಸಹಾಯ ಮಾಡುತ್ತದೆ.

ಈ ಕಾರ್ಯಕ್ಷಮತೆಯನ್ನು ಅನ್ವೇಷಿಸಲು, ನಾವು Python scikit-learn ಗ್ರಂಥಾಲಯವನ್ನು ಆಮದುಮಾಡಿ ಮತ್ತು ಪ್ರತಿಯೊಬ್ಬ ಡೇಟಾ ವಿಜ್ಞಾನಿಯೂ ನೂರಾರು ಬಾರಿ ನೋಡಿರುವ ಐಕಾನಿಕ್ ಡೇಟಾಸೆಟ್ ಅನ್ನು ಬಳಸುತ್ತೇವೆ: ಬ್ರಿಟಿಷ್ ಜೀವಶಾಸ್ತ್ರಜ್ಞ ರೋನಾಲ್ಡ್ ಫಿಶರ್ ಅವರ *Iris* ಡೇಟಾ ಸೆಟ್, 1936 ರಲ್ಲಿ ಅವರ "The use of multiple measurements in taxonomic problems" ಎಂಬ ಪತ್ರಿಕೆಯಲ್ಲಿ ಬಳಸಲಾದದು:


In [1]:
import pandas as pd
from sklearn.datasets import load_iris

iris = load_iris()
iris_df = pd.DataFrame(data=iris['data'], columns=iris['feature_names'])

### `DataFrame.shape`
ನಾವು Iris Dataset ಅನ್ನು `iris_df` ಎಂಬ ಚರದಲ್ಲಿ ಲೋಡ್ ಮಾಡಿದ್ದೇವೆ. ಡೇಟಾದೊಳಗೆ ಮುಳುಗುವ ಮೊದಲು, ನಮಗೆ ಎಷ್ಟು ಡೇಟಾಪಾಯಿಂಟ್‌ಗಳು ಇದ್ದಾರೆ ಮತ್ತು ಡೇಟಾಸೆಟ್‌ನ ಒಟ್ಟು ಗಾತ್ರ ಎಷ್ಟು ಎಂಬುದನ್ನು ತಿಳಿದುಕೊಳ್ಳುವುದು ಉಪಯುಕ್ತವಾಗಿರುತ್ತದೆ. ನಾವು ವ್ಯವಹರಿಸುತ್ತಿರುವ ಡೇಟಾದ ಪ್ರಮಾಣವನ್ನು ನೋಡುವುದು ಸಹಾಯಕವಾಗಿದೆ.


In [2]:
iris_df.shape

(150, 4)

ಹೀಗಾಗಿ, ನಾವು 150 ಸಾಲುಗಳು ಮತ್ತು 4 ಕಾಲಮ್‌ಗಳ ಡೇಟಾ ಜೊತೆ ಕೆಲಸ ಮಾಡುತ್ತಿದ್ದೇವೆ. ಪ್ರತಿ ಸಾಲು ಒಂದು ಡೇಟಾಪಾಯಿಂಟ್ ಅನ್ನು ಪ್ರತಿನಿಧಿಸುತ್ತದೆ ಮತ್ತು ಪ್ರತಿ ಕಾಲಮ್ ಡೇಟಾ ಫ್ರೇಮ್‌ಗೆ ಸಂಬಂಧಿಸಿದ ಒಂದು ವೈಶಿಷ್ಟ್ಯವನ್ನು ಪ್ರತಿನಿಧಿಸುತ್ತದೆ. ಆದ್ದರಿಂದ ಮೂಲತಃ, 4 ವೈಶಿಷ್ಟ್ಯಗಳನ್ನು ಹೊಂದಿರುವ 150 ಡೇಟಾಪಾಯಿಂಟ್‌ಗಳಿವೆ.

`shape` ಇಲ್ಲಿ ಡೇಟಾಫ್ರೇಮ್‌ನ ಒಂದು ಗುಣಲಕ್ಷಣವಾಗಿದ್ದು, ಫಂಕ್ಷನ್ ಅಲ್ಲ, ಆದ್ದರಿಂದ ಅದು ಕವಚದ ಜೋಡಿಯಲ್ಲಿ ಕೊನೆಗೊಳ್ಳುವುದಿಲ್ಲ.


### `DataFrame.columns`
ನಾವು ಈಗ 4 ಕಾಲಮ್‌ಗಳ ಡೇಟಾಗೆ ಹೋಗೋಣ. ಅವುಗಳಲ್ಲಿ ಪ್ರತಿ ಒಂದು ನಿಖರವಾಗಿ ಏನು ಪ್ರತಿನಿಧಿಸುತ್ತದೆ? `columns` ಗುಣಲಕ್ಷಣವು ಡೇಟಾಫ್ರೇಮ್‌ನ ಕಾಲಮ್‌ಗಳ ಹೆಸರನ್ನು ನಮಗೆ ನೀಡುತ್ತದೆ.


In [3]:
iris_df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')

ನಾವು ನೋಡಬಹುದು, ನಾಲ್ಕು(4) ಕಾಲಮ್‌ಗಳಿವೆ. `columns` ಗುಣಲಕ್ಷಣವು ಕಾಲಮ್‌ಗಳ ಹೆಸರನ್ನು ನಮಗೆ ತಿಳಿಸುತ್ತದೆ ಮತ್ತು ಮೂಲತಃ ಇನ್ನೇನೂ ಅಲ್ಲ. ಈ ಗುಣಲಕ್ಷಣವು ಮಹತ್ವವನ್ನು ಪಡೆಯುತ್ತದೆ ನಾವು ಡೇಟಾಸೆಟ್ ಹೊಂದಿರುವ ವೈಶಿಷ್ಟ್ಯಗಳನ್ನು ಗುರುತಿಸಲು ಬಯಸುವಾಗ.


### `DataFrame.info`
ಡೇಟಾದ ಪ್ರಮಾಣ (`shape` ಗುಣಲಕ್ಷಣದಿಂದ ನೀಡಲ್ಪಟ್ಟಿದೆ) ಮತ್ತು ವೈಶಿಷ್ಟ್ಯಗಳ ಅಥವಾ ಕಾಲಮ್‌ಗಳ ಹೆಸರುಗಳು (`columns` ಗುಣಲಕ್ಷಣದಿಂದ ನೀಡಲ್ಪಟ್ಟಿದೆ) ನಮಗೆ ಡೇಟಾಸೆಟ್ ಬಗ್ಗೆ ಏನೋ ತಿಳಿಸುತ್ತವೆ. ಈಗ, ನಾವು ಡೇಟಾಸೆಟ್‌ನಲ್ಲಿ ಇನ್ನಷ್ಟು ಆಳವಾಗಿ ಹೋಗಲು ಇಚ್ಛಿಸುತ್ತೇವೆ. `DataFrame.info()` ಫಂಕ್ಷನ್ ಇದಕ್ಕೆ ಬಹಳ ಉಪಯುಕ್ತವಾಗಿದೆ.


In [4]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


ಇಂದಿನಿಂದ, ನಾವು ಕೆಲವು ಗಮನಾರ್ಹ ಅಂಶಗಳನ್ನು ಮಾಡಬಹುದು:
1. ಪ್ರತಿ ಕಾಲಮ್‌ನ ಡೇಟಾ ಪ್ರಕಾರ: ಈ ಡೇಟಾಸೆಟ್‌ನಲ್ಲಿ, ಎಲ್ಲಾ ಡೇಟಾ 64-ಬಿಟ್ ಫ್ಲೋಟಿಂಗ್-ಪಾಯಿಂಟ್ ಸಂಖ್ಯೆಗಳಾಗಿ ಸಂಗ್ರಹಿಸಲಾಗಿದೆ.
2. ನಾನ್-ನಲ್ ಮೌಲ್ಯಗಳ ಸಂಖ್ಯೆ: ನಲ್ ಮೌಲ್ಯಗಳನ್ನು ನಿರ್ವಹಿಸುವುದು ಡೇಟಾ ತಯಾರಿಕೆಯಲ್ಲಿ ಪ್ರಮುಖ ಹಂತವಾಗಿದೆ. ಇದನ್ನು ನಂತರ ನೋಟ್ಬುಕ್‌ನಲ್ಲಿ ನಿರ್ವಹಿಸಲಾಗುವುದು.


### DataFrame.describe()
ನಮ್ಮ ಡೇಟಾಸೆಟ್‌ನಲ್ಲಿ ಬಹಳಷ್ಟು ಸಂಖ್ಯಾತ್ಮಕ ಡೇಟಾ ಇದೆ ಎಂದು ಹೇಳೋಣ. ಸರಾಸರಿ, ಮಧ್ಯಮ, ಚತುರ್ಥಾಂಶಗಳು ಇತ್ಯಾದಿ ಏಕವ್ಯತ್ಯಯ ಸಂಖ್ಯಾಶಾಸ್ತ್ರೀಯ ಲೆಕ್ಕಾಚಾರಗಳನ್ನು ಪ್ರತಿ ಕಾಲಮ್‌ಗಳ ಮೇಲೆ ಪ್ರತ್ಯೇಕವಾಗಿ ಮಾಡಬಹುದು. `DataFrame.describe()` ಫಂಕ್ಷನ್ ನಮಗೆ ಡೇಟಾಸೆಟ್‌ನ ಸಂಖ್ಯಾತ್ಮಕ ಕಾಲಮ್‌ಗಳ ಸಂಖ್ಯಾಶಾಸ್ತ್ರೀಯ ಸಾರಾಂಶವನ್ನು ಒದಗಿಸುತ್ತದೆ.


In [5]:
iris_df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
count         150.000000        150.000000         150.000000        150.000000
mean            5.843333          3.057333           3.758000          1.199333
std             0.828066          0.435866           1.765298          0.762238
min             4.300000          2.000000           1.000000          0.100000
25%             5.100000          2.800000           1.600000          0.300000
50%             5.800000          3.000000           4.350000          1.300000
75%             6.400000          3.300000           5.100000          1.800000
max             7.900000          4.400000           6.900000          2.500000

ಮೇಲಿನ ಔಟ್‌ಪುಟ್‌ನಲ್ಲಿ ಪ್ರತಿ ಕಾಲಮ್‌ನ ಒಟ್ಟು ಡೇಟಾ ಪಾಯಿಂಟ್‌ಗಳ ಸಂಖ್ಯೆ, ಸರಾಸರಿ, ಮಾನಕ ವ್ಯತ್ಯಾಸ, ಕನಿಷ್ಠ, ಕೆಳಗಿನ ಚತುರ್ಥಾಂಶ(25%), ಮಧ್ಯಮ(50%), ಮೇಲಿನ ಚತುರ್ಥಾಂಶ(75%) ಮತ್ತು ಗರಿಷ್ಠ ಮೌಲ್ಯವನ್ನು ತೋರಿಸಲಾಗಿದೆ.


### `DataFrame.head`
ಮೇಲಿನ ಎಲ್ಲಾ ಫಂಕ್ಷನ್‌ಗಳು ಮತ್ತು ಗುಣಲಕ್ಷಣಗಳೊಂದಿಗೆ, ನಮಗೆ ಡೇಟಾಸೆಟ್‌ನ ಮೇಲ್ಮಟ್ಟದ ದೃಷ್ಟಿ ಸಿಕ್ಕಿದೆ. ಎಷ್ಟು ಡೇಟಾ ಪಾಯಿಂಟ್‌ಗಳು ಇವೆ, ಎಷ್ಟು ವೈಶಿಷ್ಟ್ಯಗಳು ಇವೆ, ಪ್ರತಿ ವೈಶಿಷ್ಟ್ಯದ ಡೇಟಾ ಪ್ರಕಾರ ಮತ್ತು ಪ್ರತಿ ವೈಶಿಷ್ಟ್ಯದ ನಾನ್-ನಲ್ ಮೌಲ್ಯಗಳ ಸಂಖ್ಯೆ ನಮಗೆ ಗೊತ್ತಾಗಿದೆ.

ಈಗ ಡೇಟಾವನ್ನು ಸ್ವತಃ ನೋಡುವ ಸಮಯವಾಗಿದೆ. ನಮ್ಮ `DataFrame`ನ ಮೊದಲ ಕೆಲವು ಸಾಲುಗಳು (ಮೊದಲ ಕೆಲವು ಡೇಟಾಪಾಯಿಂಟ್‌ಗಳು) ಹೇಗಿವೆ ಎಂದು ನೋಡೋಣ:


In [6]:
iris_df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

ಇಲ್ಲಿ ಔಟ್‌ಪುಟ್ ಆಗಿ, ನಾವು ಡೇಟಾಸೆಟ್‌ನ ಐದು(5) ಎಂಟ್ರಿಗಳನ್ನು ನೋಡಬಹುದು. ಎಡಭಾಗದ ಸೂಚ್ಯಂಕವನ್ನು ನೋಡಿದರೆ, ಇವು ಮೊದಲ ಐದು ಸಾಲುಗಳು ಎಂದು ನಾವು ಕಂಡುಹಿಡಿಯಬಹುದು.


### ವ್ಯಾಯಾಮ:

ಮೇಲಿನ ಉದಾಹರಣೆಯಿಂದ ಸ್ಪಷ್ಟವಾಗುತ್ತದೆ, ಡೀಫಾಲ್ಟ್ ಆಗಿ, `DataFrame.head` ಒಂದು `DataFrame` ನ ಮೊದಲ ಐದು ಸಾಲುಗಳನ್ನು ಹಿಂತಿರುಗಿಸುತ್ತದೆ. ಕೆಳಗಿನ ಕೋಡ್ ಸೆಲ್‌ನಲ್ಲಿ, ನೀವು ಐದು ಸಾಲುಗಳಿಗಿಂತ ಹೆಚ್ಚು ಸಾಲುಗಳನ್ನು ಪ್ರದರ್ಶಿಸುವ ಮಾರ್ಗವನ್ನು ಕಂಡುಹಿಡಿಯಬಹುದೇ?


In [7]:
# Hint: Consult the documentation by using iris_df.head?

### `DataFrame.tail`
ಡೇಟಾವನ್ನು ನೋಡಲು ಇನ್ನೊಂದು ವಿಧಾನವು ಆರಂಭದ ಬದಲು ಕೊನೆಯಲ್ಲಿ ನೋಡುವುದು. `DataFrame.head` ನ ವಿರುದ್ಧವಾದದ್ದು `DataFrame.tail`, ಇದು `DataFrame` ನ ಕೊನೆಯ ಐದು ಸಾಲುಗಳನ್ನು ನೀಡುತ್ತದೆ:


In [8]:
iris_df.tail()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

ಪ್ರಾಯೋಗಿಕವಾಗಿ, ಕ್ರಮಬದ್ಧ ಡೇಟಾಸೆಟ್‌ಗಳಲ್ಲಿ ಹೊರಗಿನ ಅಂಶಗಳನ್ನು ಹುಡುಕುತ್ತಿರುವಾಗ, `DataFrame` ನ ಮೊದಲ ಕೆಲವು ಸಾಲುಗಳು ಅಥವಾ ಕೊನೆಯ ಕೆಲವು ಸಾಲುಗಳನ್ನು ಸುಲಭವಾಗಿ ಪರಿಶೀಲಿಸಲು ಸಾಧ್ಯವಾಗುವುದು ಉಪಯುಕ್ತವಾಗಿದೆ.

ಮೇಲಿನ ಎಲ್ಲಾ ಫಂಕ್ಷನ್‌ಗಳು ಮತ್ತು ಗುಣಲಕ್ಷಣಗಳು, ಕೋಡ್ ಉದಾಹರಣೆಗಳ ಸಹಾಯದಿಂದ, ನಮಗೆ ಡೇಟಾದ ಒಂದು ನೋಟ ಮತ್ತು ಅನುಭವವನ್ನು ಪಡೆಯಲು ಸಹಾಯ ಮಾಡುತ್ತವೆ.

> **ಮುಖ್ಯಾಂಶ:** DataFrame ನಲ್ಲಿ ಮಾಹಿತಿಯ ಮೆಟಾಡೇಟಾ ಅಥವಾ ಅದರ ಮೊದಲ ಮತ್ತು ಕೊನೆಯ ಕೆಲವು ಮೌಲ್ಯಗಳನ್ನು ನೋಡಿದರೆ, ನೀವು ತಕ್ಷಣವೇ ನೀವು ವ್ಯವಹರಿಸುತ್ತಿರುವ ಡೇಟಾದ ಗಾತ್ರ, ಆಕಾರ ಮತ್ತು ವಿಷಯದ ಬಗ್ಗೆ ಒಂದು ತಕ್ಷಣದ ಕಲ್ಪನೆ ಪಡೆಯಬಹುದು.


### ಕಳೆದುಹೋಗಿದ ಡೇಟಾ
ನಾವು ಕಳೆದುಹೋಗಿದ ಡೇಟಾದಲ್ಲಿ ಮುಳುಗೋಣ. ಕೆಲವು ಕಾಲಮ್‌ಗಳಲ್ಲಿ ಯಾವುದೇ ಮೌಲ್ಯ ಸಂಗ್ರಹಿಸಲಾಗದಾಗ ಕಳೆದುಹೋಗಿದ ಡೇಟಾ ಸಂಭವಿಸುತ್ತದೆ.

ಒಂದು ಉದಾಹರಣೆಯನ್ನು ತೆಗೆದುಕೊಳ್ಳೋಣ: ಯಾರಾದರೂ ತಮ್ಮ ತೂಕದ ಬಗ್ಗೆ ಜಾಗರೂಕವಾಗಿದ್ದು ಸಮೀಕ್ಷೆಯಲ್ಲಿ ತೂಕದ ಕ್ಷೇತ್ರವನ್ನು ಭರ್ತಿಮಾಡದಿದ್ದರೆ. ಆಗ ಆ ವ್ಯಕ್ತಿಯ ತೂಕ ಮೌಲ್ಯ ಕಳೆದುಹೋಗುತ್ತದೆ.

ಬಹುಮಾನವಾಗಿ, ನಿಜವಾದ ಜಗತ್ತಿನ ಡೇಟಾಸೆಟ್‌ಗಳಲ್ಲಿ ಕಳೆದುಹೋಗಿದ ಮೌಲ್ಯಗಳು ಸಂಭವಿಸುತ್ತವೆ.

**ಪಾಂಡಾಸ್ ಕಳೆದುಹೋಗಿದ ಡೇಟಾವನ್ನು ಹೇಗೆ ನಿರ್ವಹಿಸುತ್ತದೆ**


ಪಾಂಡಾಸ್ ಕಳೆದುಹೋಗಿದ ಮೌಲ್ಯಗಳನ್ನು ಎರಡು ರೀತಿಯಲ್ಲಿ ನಿರ್ವಹಿಸುತ್ತದೆ. ಮೊದಲನೆಯದು ನೀವು ಹಿಂದಿನ ವಿಭಾಗಗಳಲ್ಲಿ ನೋಡಿದ್ದೀರಿ: `NaN`, ಅಂದರೆ ನಂಬರ್ ಅಲ್ಲ. ಇದು ವಾಸ್ತವವಾಗಿ IEEE ಫ್ಲೋಟಿಂಗ್-ಪಾಯಿಂಟ್ ಸ್ಪೆಸಿಫಿಕೇಶನ್‌ನ ಭಾಗವಾಗಿರುವ ವಿಶೇಷ ಮೌಲ್ಯ ಮತ್ತು ಇದು ಕಳೆದುಹೋಗಿದ ಫ್ಲೋಟಿಂಗ್-ಪಾಯಿಂಟ್ ಮೌಲ್ಯಗಳನ್ನು ಸೂಚಿಸಲು ಮಾತ್ರ ಬಳಸಲಾಗುತ್ತದೆ.

ಫ್ಲೋಟ್ಸ್ ಹೊರತುಪಡಿಸಿ ಕಳೆದುಹೋಗಿದ ಮೌಲ್ಯಗಳಿಗೆ, ಪಾಂಡಾಸ್ ಪೈಥಾನ್ `None` ವಸ್ತುವನ್ನು ಬಳಸುತ್ತದೆ. ನೀವು ಎರಡು ವಿಭಿನ್ನ ರೀತಿಯ ಮೌಲ್ಯಗಳನ್ನು ಎದುರಿಸುವುದು ಗೊಂದಲಕಾರಿಯಾಗಬಹುದು, ಆದರೆ ಈ ವಿನ್ಯಾಸ ಆಯ್ಕೆಗೆ ತರ್ಕಬದ್ಧವಾದ ಪ್ರೋಗ್ರಾಮ್ಯಾಟಿಕ್ ಕಾರಣಗಳಿವೆ ಮತ್ತು ಪ್ರಾಯೋಗಿಕವಾಗಿ, ಈ ಮಾರ್ಗವನ್ನು ಅನುಸರಿಸುವುದರಿಂದ ಪಾಂಡಾಸ್ ಬಹುಮಟ್ಟಿನ ಪ್ರಕರಣಗಳಿಗೆ ಉತ್ತಮ ಸಮಾಧಾನವನ್ನು ನೀಡುತ್ತದೆ. ಇದನ್ನು ಬಿಟ್ಟು, `None` ಮತ್ತು `NaN` ಎರಡೂ ಬಳಸುವಾಗ ನೀವು ಜಾಗರೂಕವಾಗಿರಬೇಕಾದ ನಿರ್ಬಂಧಗಳನ್ನು ಹೊಂದಿವೆ.


### `None`: ಫ್ಲೋಟ್ ಅಲ್ಲದ ಕಾಣೆಯಾದ ಡೇಟಾ  
`None` ಪೈಥಾನ್‌ನಿಂದ ಬರುತ್ತದೆ, ಆದ್ದರಿಂದ ಅದು `'object'` ಡೇಟಾ ಪ್ರಕಾರದಲ್ಲದ NumPy ಮತ್ತು pandas ಅರೆಗಳಲ್ಲಿ ಬಳಸಲಾಗುವುದಿಲ್ಲ. ನೆನಪಿಡಿ, NumPy ಅರೆಗಳು (ಮತ್ತು pandas ನಲ್ಲಿ ಡೇಟಾ ರಚನೆಗಳು) ಒಂದೇ ಪ್ರಕಾರದ ಡೇಟಾವನ್ನು ಮಾತ್ರ ಹೊಂದಿರಬಹುದು. ಇದು ದೊಡ್ಡ ಪ್ರಮಾಣದ ಡೇಟಾ ಮತ್ತು ಗಣನಾತ್ಮಕ ಕೆಲಸಗಳಿಗೆ ಅದ್ಭುತ ಶಕ್ತಿ ನೀಡುತ್ತದೆ, ಆದರೆ ಇದರ ಲವಚಿಕತೆಯನ್ನು ಕೂಡಾ ಮಿತಿಗೊಳಿಸುತ್ತದೆ. ಇಂತಹ ಅರೆಗಳು "ಕನಿಷ್ಠ ಸಾಮಾನ್ಯ ಹಂಚಿಕೆದಾರ" ಗೆ ಅಪ್‌ಕಾಸ್ಟ್ ಮಾಡಬೇಕಾಗುತ್ತದೆ, ಅರೆದಲ್ಲಿರುವ ಎಲ್ಲವನ್ನೂ ಒಳಗೊಂಡಿರುವ ಡೇಟಾ ಪ್ರಕಾರ. ಅರೆಗಳಲ್ಲಿ `None` ಇದ್ದರೆ, ನೀವು ಪೈಥಾನ್ ವಸ್ತುಗಳೊಂದಿಗೆ ಕೆಲಸ ಮಾಡುತ್ತಿದ್ದೀರಿ ಎಂದು ಅರ್ಥ.  

ಇದನ್ನು ಕಾರ್ಯಾಚರಣೆಯಲ್ಲಿ ನೋಡಲು, ಕೆಳಗಿನ ಉದಾಹರಣೆಯ ಅರೆ (ಅದರ `dtype` ಗಮನಿಸಿ) ಪರಿಗಣಿಸಿ:


In [9]:
import numpy as np

example1 = np.array([2, None, 6, 8])
example1

array([2, None, 6, 8], dtype=object)

ಅಪ್ಕಾಸ್ಟ್ ಡೇಟಾ ಪ್ರಕಾರಗಳ ವಾಸ್ತವಿಕತೆ ಎರಡು ಪಾರ್ಶ್ವ ಪರಿಣಾಮಗಳನ್ನು ಹೊಂದಿದೆ. ಮೊದಲನೆಯದಾಗಿ, ಕಾರ್ಯಾಚರಣೆಗಳು ಸಂಯೋಜಿತ NumPy ಕೋಡ್ ಬದಲು ವ್ಯಾಖ್ಯಾನಿತ Python ಕೋಡ್ ಮಟ್ಟದಲ್ಲಿ ನಡೆಸಲಾಗುತ್ತದೆ. ಮೂಲತಃ, ಇದರಿಂದ `None` ಹೊಂದಿರುವ `Series` ಅಥವಾ `DataFrames` ಅನ್ನು ಒಳಗೊಂಡ ಯಾವುದೇ ಕಾರ್ಯಾಚರಣೆಗಳು ನಿಧಾನವಾಗುತ್ತವೆ. ನೀವು ಬಹುಶಃ ಈ ಕಾರ್ಯಕ್ಷಮತೆ ಹಾನಿಯನ್ನು ಗಮನಿಸದಿರಬಹುದು, ಆದರೆ ದೊಡ್ಡ ಡೇಟಾಸೆಟ್‌ಗಳಿಗೆ ಇದು ಸಮಸ್ಯೆಯಾಗಬಹುದು.

ಎರಡನೆಯ ಪಾರ್ಶ್ವ ಪರಿಣಾಮವು ಮೊದಲನೆಯದರಿಂದ ಉಂಟಾಗುತ್ತದೆ. ಏಕೆಂದರೆ `None` ಮೂಲತಃ `Series` ಅಥವಾ `DataFrame`ಗಳನ್ನು ವನಿಲ್ಲಾ Python ಜಗತ್ತಿಗೆ ಹಿಂದಿರುಗಿಸುತ್ತದೆ, `None` ಮೌಲ್ಯವನ್ನು ಹೊಂದಿರುವ ಅರೆಗಳ ಮೇಲೆ NumPy/pandas ಸಂಗ್ರಹಣಾ ಕಾರ್ಯಗಳು, ಉದಾಹರಣೆಗೆ `sum()` ಅಥವಾ `min()`, ಸಾಮಾನ್ಯವಾಗಿ ದೋಷವನ್ನು ಉಂಟುಮಾಡುತ್ತವೆ:


In [10]:
example1.sum()

TypeError: ignored

**ಮುಖ್ಯ ಪಾಠ**: ಪೂರ್ಣಾಂಕಗಳು ಮತ್ತು `None` ಮೌಲ್ಯಗಳ ನಡುವೆ ಸೇರಿಸುವಿಕೆ (ಮತ್ತು ಇತರ ಕಾರ್ಯಾಚರಣೆಗಳು) ನಿರ್ಧರಿಸಲಾಗಿಲ್ಲ, ಇದು ಅವುಗಳನ್ನು ಹೊಂದಿರುವ ಡೇಟಾಸೆಟ್‌ಗಳೊಂದಿಗೆ ನೀವು ಏನು ಮಾಡಬಹುದು ಎಂಬುದನ್ನು ಸೀಮಿತಗೊಳಿಸಬಹುದು.


### `NaN`: ಕಾಣೆಯಾದ ಫ್ಲೋಟ್ ಮೌಲ್ಯಗಳು

`None` ಗೆ ವಿರುದ್ಧವಾಗಿ, NumPy (ಮತ್ತು ಆದ್ದರಿಂದ pandas) ತನ್ನ ವೇಗವಾದ, ವೆಕ್ಟರೈಜ್ಡ್ ಕಾರ್ಯಾಚರಣೆಗಳು ಮತ್ತು ufuncs ಗಾಗಿ `NaN` ಅನ್ನು ಬೆಂಬಲಿಸುತ್ತದೆ. ಕೆಟ್ಟ ಸುದ್ದಿ ಎಂದರೆ `NaN` ಮೇಲೆ ನಡೆಸಲಾದ ಯಾವುದೇ ಗಣಿತವು ಯಾವಾಗಲೂ `NaN` ಅನ್ನು ಫಲಿತಾಂಶವಾಗಿ ನೀಡುತ್ತದೆ. ಉದಾಹರಣೆಗೆ:


In [11]:
np.nan + 1

nan

In [12]:
np.nan * 0

nan

ಚೆನ್ನಾದ ಸುದ್ದಿ: `NaN` ಇರುವ ಅರೆಗಳ ಮೇಲೆ ಸಂಗ್ರಹಣೆಗಳು ದೋಷಗಳನ್ನು ತೋರಿಸುವುದಿಲ್ಲ. ಕೆಟ್ಟ ಸುದ್ದಿ: ಫಲಿತಾಂಶಗಳು ಸಮಾನವಾಗಿ ಉಪಯುಕ್ತವಾಗಿರುವುದಿಲ್ಲ:


In [13]:
example2 = np.array([2, np.nan, 6, 8]) 
example2.sum(), example2.min(), example2.max()

(nan, nan, nan)

### ವ್ಯಾಯಾಮ:


In [11]:
# What happens if you add np.nan and None together?


ಸ್ಮರಿಸಿ: `NaN` ಕೇವಲ ಕಾಣೆಯಾದ ಫ್ಲೋಟಿಂಗ್-ಪಾಯಿಂಟ್ ಮೌಲ್ಯಗಳಿಗಾಗಿ ಮಾತ್ರ; ಪೂರ್ಣಾಂಕಗಳು, ಸ್ಟ್ರಿಂಗ್‌ಗಳು ಅಥವಾ ಬೂಲಿಯನ್‌ಗಳಿಗೆ `NaN` ಸಮಾನಾರ್ಥಕವಿಲ್ಲ.


### `NaN` ಮತ್ತು `None`: pandas ನಲ್ಲಿ ಶೂನ್ಯ ಮೌಲ್ಯಗಳು

`NaN` ಮತ್ತು `None` ಸ್ವಲ್ಪ ವಿಭಿನ್ನವಾಗಿ ವರ್ತಿಸಬಹುದು ಆದರೂ, pandas ಅವುಗಳನ್ನು ಪರಸ್ಪರ ಬದಲಾಯಿಸಬಹುದಾದಂತೆ ನಿರ್ವಹಿಸಲು ನಿರ್ಮಿಸಲಾಗಿದೆ. ನಾವು ಏನು ಅರ್ಥಮಾಡಿಕೊಳ್ಳುತ್ತೇವೆ ಎಂದು ನೋಡಲು, ಪೂರ್ಣಾಂಕಗಳ `Series` ಅನ್ನು ಪರಿಗಣಿಸಿ:


In [15]:
int_series = pd.Series([1, 2, 3], dtype=int)
int_series

0    1
1    2
2    3
dtype: int64

### ವ್ಯಾಯಾಮ:


In [16]:
# Now set an element of int_series equal to None.
# How does that element show up in the Series?
# What is the dtype of the Series?


`Series` ಮತ್ತು `DataFrame`ಗಳಲ್ಲಿ ಡೇಟಾ ಸಮಾನತೆಯನ್ನು ಸ್ಥಾಪಿಸಲು ಡೇಟಾ ಪ್ರಕಾರಗಳನ್ನು ಅಪ್‌ಕಾಸ್ಟ್ ಮಾಡುವ ಪ್ರಕ್ರಿಯೆಯಲ್ಲಿ, pandas `None` ಮತ್ತು `NaN` ನಡುವಿನ ಕಾಣೆಯ ಮೌಲ್ಯಗಳನ್ನು ಸ್ವೀಕರಿಸಲು ಇಚ್ಛಿಸುತ್ತದೆ. ಈ ವಿನ್ಯಾಸ ವೈಶಿಷ್ಟ್ಯದ ಕಾರಣದಿಂದ, pandas ನಲ್ಲಿ `None` ಮತ್ತು `NaN` ಅನ್ನು "null" ಎಂಬ ಎರಡು ವಿಭಿನ್ನ ರುಚಿಗಳಾಗಿ ಪರಿಗಣಿಸುವುದು ಸಹಾಯಕವಾಗಬಹುದು. ನಿಜವಾಗಿಯೂ, pandas ನಲ್ಲಿ ಕಾಣೆಯ ಮೌಲ್ಯಗಳನ್ನು ನಿರ್ವಹಿಸಲು ನೀವು ಬಳಸುವ ಕೆಲವು ಮೂಲ ವಿಧಾನಗಳು ಈ ಕಲ್ಪನೆಯನ್ನು ಅವರ ಹೆಸರಿನಲ್ಲಿ ಪ್ರತಿಬಿಂಬಿಸುತ್ತವೆ:

- `isnull()`: ಕಾಣೆಯ ಮೌಲ್ಯಗಳನ್ನು ಸೂಚಿಸುವ ಬೂಲಿಯನ್ ಮಾಸ್ಕ್ ಅನ್ನು ರಚಿಸುತ್ತದೆ
- `notnull()`: `isnull()` ಗೆ ವಿರುದ್ಧ
- `dropna()`: ಡೇಟಾದ ಫಿಲ್ಟರ್ ಮಾಡಲಾದ ಆವೃತ್ತಿಯನ್ನು ಹಿಂತಿರುಗಿಸುತ್ತದೆ
- `fillna()`: ಕಾಣೆಯ ಮೌಲ್ಯಗಳನ್ನು ತುಂಬಿದ ಅಥವಾ ಅಂದಾಜಿಸಿದ ಡೇಟಾದ ನಕಲನ್ನು ಹಿಂತಿರುಗಿಸುತ್ತದೆ

ಇವುಗಳನ್ನು ನಿಪುಣವಾಗಿ ಬಳಸುವುದು ಮತ್ತು ಆರಾಮವಾಗಿ ತಿಳಿದುಕೊಳ್ಳುವುದು ಮುಖ್ಯ, ಆದ್ದರಿಂದ ನಾವು ಪ್ರತಿಯೊಂದರನ್ನೂ ಕೆಲವು ಆಳದಲ್ಲಿ ಪರಿಶೀಲಿಸೋಣ.


### ಶೂನ್ಯ ಮೌಲ್ಯಗಳನ್ನು ಪತ್ತೆಹಚ್ಚುವುದು

ನಾವು ಕಳೆದುಹೋಗಿರುವ ಮೌಲ್ಯಗಳ ಮಹತ್ವವನ್ನು ಅರ್ಥಮಾಡಿಕೊಂಡಿದ್ದೇವೆ, ಈಗ ಅವುಗಳನ್ನು ನಮ್ಮ ಡೇಟಾಸೆಟ್‌ನಲ್ಲಿ ಪತ್ತೆಹಚ್ಚಬೇಕಾಗಿದೆ, ಅವುಗಳೊಂದಿಗೆ ವ್ಯವಹರಿಸುವ ಮೊದಲು.  
`isnull()` ಮತ್ತು `notnull()` ಎರಡೂ ನಿಮ್ಮ ಪ್ರಾಥಮಿಕ ವಿಧಾನಗಳು ಶೂನ್ಯ ಡೇಟಾವನ್ನು ಪತ್ತೆಹಚ್ಚಲು. ಎರಡೂ ನಿಮ್ಮ ಡೇಟಾದ ಮೇಲೆ ಬೂಲಿಯನ್ ಮಾಸ್ಕ್‌ಗಳನ್ನು ಹಿಂತಿರುಗಿಸುತ್ತವೆ.


In [17]:
example3 = pd.Series([0, np.nan, '', None])

In [18]:
example3.isnull()

0    False
1     True
2    False
3     True
dtype: bool

ಔಟ್‌ಪುಟ್ ಅನ್ನು ನಿಕಟವಾಗಿ ನೋಡಿ. ಇದರಲ್ಲಿ ಯಾವುದಾದರೂ ನಿಮಗೆ ಆಶ್ಚರ್ಯಕರವಾಗಿದೆಯೇ? `0` ಗಣಿತೀಯ ಶೂನ್ಯವಾಗಿದ್ದರೂ, ಅದು ಸಂಪೂರ್ಣವಾಗಿ ಒಳ್ಳೆಯ ಪೂರ್ಣಾಂಕವಾಗಿದೆ ಮತ್ತು pandas ಅದನ್ನು ಹಾಗೆಯೇ ಪರಿಗಣಿಸುತ್ತದೆ. `''` ಸ್ವಲ್ಪ ಹೆಚ್ಚು ಸೂಕ್ಷ್ಮವಾಗಿದೆ. ನಾವು ಅದನ್ನು ವಿಭಾಗ 1 ರಲ್ಲಿ ಖಾಲಿ ಸ್ಟ್ರಿಂಗ್ ಮೌಲ್ಯವನ್ನು ಪ್ರತಿನಿಧಿಸಲು ಬಳಸಿದ್ದರೂ, ಅದು pandas ಗೆ ಸಂಬಂಧಿಸಿದಂತೆ ಶೂನ್ಯದ ಪ್ರತಿನಿಧನೆ ಅಲ್ಲ, ಅದು ಸ್ಟ್ರಿಂಗ್ ವಸ್ತುವಾಗಿದೆ.

ಈಗ, ಇದನ್ನು ತಿರುಗಿಸಿ, ನೀವು ಪ್ರಾಯೋಗಿಕವಾಗಿ ಬಳಸುವ ರೀತಿಯಲ್ಲಿ ಈ ವಿಧಾನಗಳನ್ನು ಬಳಸೋಣ. ನೀವು Boolean ಮಾಸ್ಕ್‌ಗಳನ್ನು ನೇರವಾಗಿ ``Series`` ಅಥವಾ ``DataFrame`` ಸೂಚ್ಯಂಕವಾಗಿ ಬಳಸಬಹುದು, ಇದು ಪ್ರತ್ಯೇಕವಾಗಿ ಕಾಣೆಯಾದ (ಅಥವಾ ಇರುವ) ಮೌಲ್ಯಗಳೊಂದಿಗೆ ಕೆಲಸ ಮಾಡಲು ಉಪಯುಕ್ತವಾಗಬಹುದು.

ನಾವು ಒಟ್ಟು ಕಾಣೆಯಾದ ಮೌಲ್ಯಗಳ ಸಂಖ್ಯೆಯನ್ನು ತಿಳಿದುಕೊಳ್ಳಬೇಕಾದರೆ, `isnull()` ವಿಧಾನದಿಂದ ಉತ್ಪನ್ನವಾದ ಮಾಸ್ಕ್ ಮೇಲೆ ಸರಳವಾಗಿ ಮೊತ್ತವನ್ನು ಮಾಡಬಹುದು.


In [19]:
example3.isnull().sum()

2

### ವ್ಯಾಯಾಮ:


In [20]:
# Try running example3[example3.notnull()].
# Before you do so, what do you expect to see?


**ಮುಖ್ಯ ಅಂಶ**: `isnull()` ಮತ್ತು `notnull()` ವಿಧಾನಗಳು DataFrames ನಲ್ಲಿ ಬಳಸಿದಾಗ ಸಮಾನ ಫಲಿತಾಂಶಗಳನ್ನು ನೀಡುತ್ತವೆ: ಅವು ಫಲಿತಾಂಶಗಳನ್ನು ಮತ್ತು ಆ ಫಲಿತಾಂಶಗಳ ಸೂಚ್ಯಂಕವನ್ನು ತೋರಿಸುತ್ತವೆ, ಇದು ನಿಮ್ಮ ಡೇಟಾ ಜೊತೆ ಹೋರಾಡುವಾಗ ನಿಮಗೆ ಬಹಳ ಸಹಾಯ ಮಾಡುತ್ತದೆ.


### ಕಾಣೆಯಾದ ಡೇಟಾವನ್ನು ನಿಭಾಯಿಸುವುದು

> **ಕಲಿಕೆಯ ಗುರಿ:** ಈ ಉಪವಿಭಾಗದ ಕೊನೆಯಲ್ಲಿ, ನೀವು DataFrames ನಿಂದ ನಲ್ ಮೌಲ್ಯಗಳನ್ನು ಹೇಗೆ ಮತ್ತು ಯಾವಾಗ ಬದಲಾಯಿಸಬೇಕು ಅಥವಾ ತೆಗೆದುಹಾಕಬೇಕು ಎಂಬುದನ್ನು ತಿಳಿದುಕೊಳ್ಳಬೇಕು.

ಯಂತ್ರ ಅಧ್ಯಯನ ಮಾದರಿಗಳು ತಾವು ಕಾಣೆಯಾದ ಡೇಟಾವನ್ನು ನಿಭಾಯಿಸಲು ಸಾಧ್ಯವಿಲ್ಲ. ಆದ್ದರಿಂದ, ಡೇಟಾವನ್ನು ಮಾದರಿಯಲ್ಲಿ ಹಂಚಿಕೆಯಾಗಿಸುವ ಮೊದಲು, ನಾವು ಈ ಕಾಣೆಯಾದ ಮೌಲ್ಯಗಳನ್ನು ನಿಭಾಯಿಸಬೇಕಾಗುತ್ತದೆ.

ಕಾಣೆಯಾದ ಡೇಟಾವನ್ನು ಹೇಗೆ ನಿಭಾಯಿಸಲಾಗುತ್ತದೆ ಎಂಬುದು ಸೂಕ್ಷ್ಮ ವ್ಯವಹಾರಗಳನ್ನು ಹೊಂದಿದೆ, ಇದು ನಿಮ್ಮ ಅಂತಿಮ ವಿಶ್ಲೇಷಣೆ ಮತ್ತು ನೈಜ ಜಗತ್ತಿನ ಫಲಿತಾಂಶಗಳನ್ನು ಪ್ರಭಾವಿಸುತ್ತದೆ.

ಕಾಣೆಯಾದ ಡೇಟಾವನ್ನು ನಿಭಾಯಿಸುವ ಎರಡು ಪ್ರಮುಖ ವಿಧಾನಗಳಿವೆ:


1.   ಕಾಣೆಯಾದ ಮೌಲ್ಯವನ್ನು ಹೊಂದಿರುವ ಸಾಲನ್ನು ತೆಗೆದುಹಾಕುವುದು
2.   ಕಾಣೆಯಾದ ಮೌಲ್ಯವನ್ನು ಬೇರೆ ಮೌಲ್ಯದಿಂದ ಬದಲಾಯಿಸುವುದು

ನಾವು ಈ ಎರಡೂ ವಿಧಾನಗಳನ್ನು ಮತ್ತು ಅವುಗಳ ಲಾಭ-ನಷ್ಟಗಳನ್ನು ವಿವರವಾಗಿ ಚರ್ಚಿಸುವೆವು.


### ನಲ್ ಮೌಲ್ಯಗಳನ್ನು ಬಿಟ್ಟಿಹಾಕುವುದು

ನಾವು ನಮ್ಮ ಮಾದರಿಗೆ ನೀಡುವ ಡೇಟಾದ ಪ್ರಮಾಣವು ಅದರ ಕಾರ್ಯಕ್ಷಮತೆಯ ಮೇಲೆ ನೇರ ಪರಿಣಾಮ ಬೀರುತ್ತದೆ. ನಲ್ ಮೌಲ್ಯಗಳನ್ನು ಬಿಟ್ಟಿಹಾಕುವುದು ಎಂದರೆ ನಾವು ಡೇಟಾಪಾಯಿಂಟ್‌ಗಳ ಸಂಖ್ಯೆಯನ್ನು ಕಡಿಮೆ ಮಾಡುತ್ತಿದ್ದೇವೆ, ಮತ್ತು ಆದ್ದರಿಂದ ಡೇಟಾಸೆಟ್‌ನ ಗಾತ್ರವನ್ನು ಕಡಿಮೆ ಮಾಡುತ್ತಿದ್ದೇವೆ. ಆದ್ದರಿಂದ, ಡೇಟಾಸೆಟ್ ಬಹಳ ದೊಡ್ಡದಾಗಿದ್ದಾಗ ನಲ್ ಮೌಲ್ಯಗಳಿರುವ ಸಾಲುಗಳನ್ನು ಬಿಟ್ಟಿಹಾಕುವುದು ಶಿಫಾರಸು ಮಾಡಲಾಗುತ್ತದೆ.

ಮತ್ತೊಂದು ಉದಾಹರಣೆ ಎಂದರೆ ಒಂದು ನಿರ್ದಿಷ್ಟ ಸಾಲು ಅಥವಾ ಕಾಲಮ್‌ನಲ್ಲಿ ಬಹಳಷ್ಟು ಮಿಸ್ ಆಗಿರುವ ಮೌಲ್ಯಗಳಿರಬಹುದು. ಆಗ ಅವುಗಳನ್ನು ಬಿಟ್ಟಿಹಾಕಬಹುದು ಏಕೆಂದರೆ ಆ ಸಾಲು/ಕಾಲಮ್‌ಗೆ ಹೆಚ್ಚಿನ ಡೇಟಾ ಇಲ್ಲದಿರುವುದರಿಂದ ಅವು ನಮ್ಮ ವಿಶ್ಲೇಷಣೆಗೆ ಹೆಚ್ಚಿನ ಮೌಲ್ಯವನ್ನು ಸೇರಿಸುವುದಿಲ್ಲ.

ಮಿಸ್ ಆಗಿರುವ ಮೌಲ್ಯಗಳನ್ನು ಗುರುತಿಸುವುದನ್ನು ಮೀರಿ, pandas `Series` ಮತ್ತು `DataFrame`ಗಳಿಂದ ನಲ್ ಮೌಲ್ಯಗಳನ್ನು ತೆಗೆದುಹಾಕಲು ಅನುಕೂಲಕರ ವಿಧಾನವನ್ನು ಒದಗಿಸುತ್ತದೆ. ಇದನ್ನು ಕಾರ್ಯದಲ್ಲಿ ನೋಡಲು, ನಾವು `example3` ಗೆ ಮರಳೋಣ. `DataFrame.dropna()` ಫಂಕ್ಷನ್ ನಲ್ ಮೌಲ್ಯಗಳಿರುವ ಸಾಲುಗಳನ್ನು ಬಿಟ್ಟಿಹಾಕಲು ಸಹಾಯ ಮಾಡುತ್ತದೆ.


In [21]:
example3 = example3.dropna()
example3

0    0
2     
dtype: object

ನೋಟ್ ಮಾಡಿ ಇದು ನಿಮ್ಮ `example3[example3.notnull()]` ನಿಂದ ನಿಮ್ಮ ಔಟ್‌ಪುಟ್‌ನಂತೆ ಕಾಣಬೇಕು. ಇಲ್ಲಿ ವ್ಯತ್ಯಾಸವೆಂದರೆ, ಮಸ್ಕ್ ಮಾಡಲಾದ ಮೌಲ್ಯಗಳ ಮೇಲೆ ಮಾತ್ರ ಸೂಚ್ಯಂಕ ಹಾಕುವುದರ ಬದಲು, `dropna` ಆ ಕಳೆದುಹೋಗಿದ ಮೌಲ್ಯಗಳನ್ನು `Series` `example3` ನಿಂದ ತೆಗೆದುಹಾಕಿದೆ.

ಕಾರಣ DataFrames ಗೆ ಎರಡು ಆಯಾಮಗಳಿವೆ, ಅವು ಡೇಟಾವನ್ನು ತೆಗೆದುಹಾಕಲು ಹೆಚ್ಚು ಆಯ್ಕೆಗಳನ್ನು ಒದಗಿಸುತ್ತವೆ.


In [22]:
example4 = pd.DataFrame([[1,      np.nan, 7], 
                         [2,      5,      8], 
                         [np.nan, 6,      9]])
example4

0    1  2
0  1.0  NaN  7
1  2.0  5.0  8
2  NaN  6.0  9

(ನೀವು ಗಮನಿಸಿದ್ದೀರಾ pandas ಎರಡು ಕಾಲಮ್‌ಗಳನ್ನು `NaN` ಗಳನ್ನು ಹೊಂದಿಸಲು ಫ್ಲೋಟ್‌ಗಳಿಗೆ ಅಪ್‌ಕಾಸ್ಟ್ ಮಾಡುತ್ತದೆ?)

ನೀವು `DataFrame` ನಿಂದ ಒಂದೇ ಮೌಲ್ಯವನ್ನು ತೆಗೆದುಹಾಕಲು ಸಾಧ್ಯವಿಲ್ಲ, ಆದ್ದರಿಂದ ನೀವು ಸಂಪೂರ್ಣ ಸಾಲುಗಳು ಅಥವಾ ಕಾಲಮ್‌ಗಳನ್ನು ತೆಗೆದುಹಾಕಬೇಕು. ನೀವು ಏನು ಮಾಡುತ್ತಿದ್ದೀರೋ ಅದಕ್ಕೆ ಅನುಗುಣವಾಗಿ, ನೀವು ಒಂದನ್ನು ಅಥವಾ ಇನ್ನೊಂದನ್ನು ಮಾಡಬಹುದು, ಹಾಗಾಗಿ pandas ಎರಡಕ್ಕೂ ಆಯ್ಕೆಗಳು ನೀಡುತ್ತದೆ. ಡೇಟಾ ಸೈನ್ಸ್‌ನಲ್ಲಿ, ಕಾಲಮ್‌ಗಳು ಸಾಮಾನ್ಯವಾಗಿ ಚರಗಳನ್ನು ಪ್ರತಿನಿಧಿಸುತ್ತವೆ ಮತ್ತು ಸಾಲುಗಳು ಅವಲೋಕನಗಳನ್ನು ಪ್ರತಿನಿಧಿಸುತ್ತವೆ, ಆದ್ದರಿಂದ ನೀವು ಡೇಟಾದ ಸಾಲುಗಳನ್ನು ತೆಗೆದುಹಾಕುವ ಸಾಧ್ಯತೆ ಹೆಚ್ಚು; `dropna()` ನ ಡೀಫಾಲ್ಟ್ ಸೆಟ್ಟಿಂಗ್ ಯಾವುದೇ ನಲ್ ಮೌಲ್ಯಗಳನ್ನು ಹೊಂದಿರುವ ಎಲ್ಲಾ ಸಾಲುಗಳನ್ನು ತೆಗೆದುಹಾಕುವದು:


In [23]:
example4.dropna()

0    1  2
1  2.0  5.0  8

ಅಗತ್ಯವಿದ್ದರೆ, ನೀವು ಕಾಲಮ್‌ಗಳಿಂದ NA ಮೌಲ್ಯಗಳನ್ನು ತೆಗೆದುಹಾಕಬಹುದು. ಅದಕ್ಕಾಗಿ `axis=1` ಅನ್ನು ಬಳಸಿ:


In [24]:
example4.dropna(axis='columns')

2
0  7
1  8
2  9

ಗಮನಿಸಿ, ಇದು ನೀವು ಉಳಿಸಿಕೊಳ್ಳಲು ಬಯಸುವ ಬಹಳಷ್ಟು ಡೇಟಾವನ್ನು ಬಿಟ್ಟಿಹೋಗಬಹುದು, ವಿಶೇಷವಾಗಿ ಸಣ್ಣ ಡೇಟಾಸೆಟ್‌ಗಳಲ್ಲಿ. ನೀವು ಕೆಲವೊಂದು ಅಥವಾ ಎಲ್ಲಾ ನಲ್ ಮೌಲ್ಯಗಳನ್ನು ಹೊಂದಿರುವ ಸಾಲುಗಳು ಅಥವಾ ಕಾಲಮ್‌ಗಳನ್ನು ಮಾತ್ರ ಬಿಟ್ಟಿಹೋಗಲು ಬಯಸಿದರೆ ಏನು ಮಾಡಬೇಕು? ನೀವು `dropna` ನಲ್ಲಿ `how` ಮತ್ತು `thresh` ಪರಿಮಾಣಗಳನ್ನು ಬಳಸಿಕೊಂಡು ಆ ಸೆಟ್ಟಿಂಗ್‌ಗಳನ್ನು ನಿರ್ದಿಷ್ಟಪಡಿಸಬಹುದು.

ಡೀಫಾಲ್ಟ್ ಆಗಿ, `how='any'` (ನೀವು ಸ್ವತಃ ಪರಿಶೀಲಿಸಲು ಅಥವಾ ಈ ವಿಧಾನದಲ್ಲಿ ಇನ್ನೇನು ಪರಿಮಾಣಗಳಿವೆ ಎಂದು ನೋಡಲು ಬಯಸಿದರೆ, ಕೋಡ್ ಸೆಲ್‌ನಲ್ಲಿ `example4.dropna?` ಅನ್ನು ರನ್ ಮಾಡಿ). ಬದಲಾಗಿ, ನೀವು `how='all'` ಅನ್ನು ನಿರ್ದಿಷ್ಟಪಡಿಸಬಹುದು, ಇದರಿಂದ ಎಲ್ಲಾ ನಲ್ ಮೌಲ್ಯಗಳನ್ನು ಹೊಂದಿರುವ ಸಾಲುಗಳು ಅಥವಾ ಕಾಲಮ್‌ಗಳನ್ನು ಮಾತ್ರ ಬಿಟ್ಟಿಹೋಗುತ್ತದೆ. ಮುಂದಿನ ವ್ಯಾಯಾಮದಲ್ಲಿ ಇದನ್ನು ಕಾರ್ಯಾಚರಣೆಯಲ್ಲಿ ನೋಡಲು ನಮ್ಮ ಉದಾಹರಣೆಯ `DataFrame` ಅನ್ನು ವಿಸ್ತರಿಸೋಣ.


In [25]:
example4[3] = np.nan
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

> ಮುಖ್ಯ ಅಂಶಗಳು:  
1. ಡೇಟಾಸೆಟ್ ಸಾಕಷ್ಟು ದೊಡ್ಡದಾಗಿದ್ದರೆ ಮಾತ್ರ ನಲ್ ಮೌಲ್ಯಗಳನ್ನು ಬಿಟ್ಟಿಹಾಕುವುದು ಒಳ್ಳೆಯ ವಿಚಾರ.  
2. ಹೆಚ್ಚಿನ ಡೇಟಾ ಇಲ್ಲದಿದ್ದರೆ ಸಂಪೂರ್ಣ ಸಾಲುಗಳು ಅಥವಾ ಕಾಲಮ್‌ಗಳನ್ನು ಬಿಟ್ಟಿಹಾಕಬಹುದು.  
3. `DataFrame.dropna(axis=)` ವಿಧಾನವು ನಲ್ ಮೌಲ್ಯಗಳನ್ನು ಬಿಟ್ಟಿಹಾಕಲು ಸಹಾಯ ಮಾಡುತ್ತದೆ. `axis` ಆರ್ಗ್ಯುಮೆಂಟ್ ಸಾಲುಗಳನ್ನು ಬಿಟ್ಟಿಹಾಕಬೇಕೇ ಅಥವಾ ಕಾಲಮ್‌ಗಳನ್ನು ಬಿಟ್ಟಿಹಾಕಬೇಕೇ ಎಂಬುದನ್ನು ಸೂಚಿಸುತ್ತದೆ.  
4. `how` ಆರ್ಗ್ಯುಮೆಂಟ್ ಕೂಡ ಬಳಸಬಹುದು. ಡೀಫಾಲ್ಟ್ ಆಗಿ ಇದು `any` ಗೆ ಸೆಟ್ ಆಗಿರುತ್ತದೆ. ಆದ್ದರಿಂದ, ಯಾವುದೇ ನಲ್ ಮೌಲ್ಯಗಳನ್ನು ಹೊಂದಿರುವ ಸಾಲುಗಳು/ಕಾಲಮ್‌ಗಳನ್ನು ಮಾತ್ರ ಬಿಟ್ಟಿಹಾಕುತ್ತದೆ. ಎಲ್ಲಾ ಮೌಲ್ಯಗಳು ನಲ್ ಆಗಿರುವ ಸಾಲುಗಳು/ಕಾಲಮ್‌ಗಳನ್ನು ಮಾತ್ರ ಬಿಟ್ಟಿಹಾಕಲು ಇದನ್ನು `all` ಗೆ ಸೆಟ್ ಮಾಡಬಹುದು.


### ವ್ಯಾಯಾಮ:


In [22]:
# How might you go about dropping just column 3?
# Hint: remember that you will need to supply both the axis parameter and the how parameter.


`thresh` ಪರಿಮಾಣವು ನಿಮಗೆ ಸೂಕ್ಷ್ಮ ನಿಯಂತ್ರಣವನ್ನು ನೀಡುತ್ತದೆ: ನೀವು ಸಾಲು ಅಥವಾ ಕಾಲಮ್ ಉಳಿಸಿಕೊಳ್ಳಲು ಅಗತ್ಯವಿರುವ *ನಲ್ ಅಲ್ಲದ* ಮೌಲ್ಯಗಳ ಸಂಖ್ಯೆಯನ್ನು ಹೊಂದಿಸುತ್ತೀರಿ:


In [27]:
example4.dropna(axis='rows', thresh=3)

0    1  2   3
1  2.0  5.0  8 NaN

ಇಲ್ಲಿ, ಮೊದಲ ಮತ್ತು ಕೊನೆಯ ಸಾಲುಗಳನ್ನು ತೆಗೆದುಹಾಕಲಾಗಿದೆ, ಏಕೆಂದರೆ ಅವುಗಳಲ್ಲಿ ಕೇವಲ ಎರಡು ಅಶೂನ್ಯ ಮೌಲ್ಯಗಳಿವೆ.


### ಶೂನ್ಯ ಮೌಲ್ಯಗಳನ್ನು ತುಂಬುವುದು

ಕಾಲಕಾಲಕ್ಕೆ, ಶೂನ್ಯ ಮೌಲ್ಯಗಳನ್ನು ಮಾನ್ಯವಾಗಬಹುದಾದ ಮೌಲ್ಯಗಳಿಂದ ತುಂಬುವುದು ಅರ್ಥಪೂರ್ಣವಾಗಬಹುದು. ಶೂನ್ಯ ಮೌಲ್ಯಗಳನ್ನು ತುಂಬಲು ಕೆಲವು ತಂತ್ರಗಳು ಇವೆ. ಮೊದಲನೆಯದು ಡೊಮೇನ್ ಜ್ಞಾನವನ್ನು (ಡೇಟಾಸೆಟ್ ಆಧಾರಿತ ವಿಷಯದ ಜ್ಞಾನ) ಬಳಸಿಕೊಂಡು ಕೇವಲ ಶೂನ್ಯ ಮೌಲ್ಯಗಳನ್ನು ಅಂದಾಜಿಸುವುದು.

ನೀವು ಇದನ್ನು ಸ್ಥಳದಲ್ಲಿಯೇ ಮಾಡಲು `isnull` ಅನ್ನು ಬಳಸಬಹುದು, ಆದರೆ ತುಂಬಾ ಮೌಲ್ಯಗಳನ್ನು ತುಂಬಬೇಕಾದರೆ ಅದು ಕಷ್ಟಕರವಾಗಬಹುದು. ಡೇಟಾ ಸೈನ್ಸ್‌ನಲ್ಲಿ ಇದು ಸಾಮಾನ್ಯ ಕಾರ್ಯವಾಗಿರುವುದರಿಂದ, pandas `fillna` ಅನ್ನು ಒದಗಿಸುತ್ತದೆ, ಇದು `Series` ಅಥವಾ `DataFrame` ನ ಪ್ರತಿಯನ್ನು ಹಿಂತಿರುಗಿಸುತ್ತದೆ, ಅಲ್ಲಿ ಶೂನ್ಯ ಮೌಲ್ಯಗಳನ್ನು ನೀವು ಆಯ್ಕೆಮಾಡಿದ ಮೌಲ್ಯದಿಂದ ಬದಲಿಸಲಾಗಿದೆ. ಇದನ್ನು ಪ್ರಾಯೋಗಿಕವಾಗಿ ಹೇಗೆ ಕೆಲಸ ಮಾಡುತ್ತದೆ ಎಂದು ನೋಡಲು ಮತ್ತೊಂದು ಉದಾಹರಣೆಯ `Series` ಅನ್ನು ರಚಿಸೋಣ.


### ವರ್ಗೀಕೃತ ಡೇಟಾ (ಸಂಖ್ಯಾತ್ಮಕವಲ್ಲದ)
ಮೊದಲು ನಾವು ಸಂಖ್ಯಾತ್ಮಕವಲ್ಲದ ಡೇಟಾವನ್ನು ಪರಿಗಣಿಸೋಣ. ಡೇಟಾಸೆಟ್‌ಗಳಲ್ಲಿ, ನಮಗೆ ವರ್ಗೀಕೃತ ಡೇಟಾ ಇರುವ ಕಾಲಮ್‌ಗಳು ಇರುತ್ತವೆ. ಉದಾ. ಲಿಂಗ, ಸತ್ಯ ಅಥವಾ ಅಸತ್ಯ ಇತ್ಯಾದಿ.

ಇವುಗಳಲ್ಲಿ ಬಹುತೇಕ ಸಂದರ್ಭಗಳಲ್ಲಿ, ನಾವು ಕಳವಳಾದ ಮೌಲ್ಯಗಳನ್ನು ಕಾಲಮ್‌ನ `mode` (ಅತ್ಯಧಿಕ ಸಂಭವನೀಯ ಮೌಲ್ಯ) ಮೂಲಕ ಬದಲಾಯಿಸುತ್ತೇವೆ. ಉದಾಹರಣೆಗೆ, ನಮಗೆ 100 ಡೇಟಾ ಪಾಯಿಂಟ್‌ಗಳಿದ್ದು, 90 ಸತ್ಯ ಎಂದು ಹೇಳಿದ್ದು, 8 ಅಸತ್ಯ ಎಂದು ಹೇಳಿದ್ದು ಮತ್ತು 2 ಭರ್ತಿ ಮಾಡಿಲ್ಲ. ಆಗ, ನಾವು ಆ 2 ಅನ್ನು ಸಂಪೂರ್ಣ ಕಾಲಮ್ ಪರಿಗಣಿಸಿ ಸತ್ಯದಿಂದ ಭರ್ತಿ ಮಾಡಬಹುದು.

ಮತ್ತೆ, ಇಲ್ಲಿ ನಾವು ಕ್ಷೇತ್ರ ಜ್ಞಾನವನ್ನು ಬಳಸಬಹುದು. ಮೋಡ್‌ನೊಂದಿಗೆ ಭರ್ತಿ ಮಾಡುವ ಉದಾಹರಣೆಯನ್ನು ಪರಿಗಣಿಸೋಣ.


In [28]:
fill_with_mode = pd.DataFrame([[1,2,"True"],
                               [3,4,None],
                               [5,6,"False"],
                               [7,8,"True"],
                               [9,10,"True"]])

fill_with_mode

0   1      2
0  1   2   True
1  3   4   None
2  5   6  False
3  7   8   True
4  9  10   True

ಈಗ, ಮೊದಲು `None` ಮೌಲ್ಯವನ್ನು ಮೋಡ್‌ನೊಂದಿಗೆ ತುಂಬಿಸುವ ಮೊದಲು ಮೋಡ್ ಅನ್ನು ಕಂಡುಹಿಡಿಯೋಣ.


In [29]:
fill_with_mode[2].value_counts()

True     3
False    1
Name: 2, dtype: int64

ಹೀಗಾಗಿ, ನಾವು None ಅನ್ನು True ಮೂಲಕ ಬದಲಾಯಿಸುವೆವು


In [30]:
fill_with_mode[2].fillna('True',inplace=True)

In [31]:
fill_with_mode

0   1      2
0  1   2   True
1  3   4   True
2  5   6  False
3  7   8   True
4  9  10   True

ನಾವು ನೋಡಬಹುದು, ನಲ್ ಮೌಲ್ಯವನ್ನು ಬದಲಾಯಿಸಲಾಗಿದೆ. ಹೇಳಬೇಕಾಗಿಲ್ಲ, ನಾವು ಯಾವುದೇ ವಾಕ್ಯವನ್ನು `'True'` ಬದಲು ಬರೆಯಬಹುದು ಮತ್ತು ಅದು ಬದಲಾಯಿಸಲಾಗುತ್ತಿತ್ತು.


### ಸಂಖ್ಯಾತ್ಮಕ ಡೇಟಾ
ಈಗ, ಸಂಖ್ಯಾತ್ಮಕ ಡೇಟಾಕ್ಕೆ ಬನ್ನೋಣ. ಇಲ್ಲಿ, ನಾವು ಕಾಣುವ ಎರಡು ಸಾಮಾನ್ಯ ವಿಧಾನಗಳು ಇಲ್ಲಿವೆ:

1. ಸಾಲಿನ ಮಧ್ಯಮ ಮೌಲ್ಯದಿಂದ ಬದಲಾಯಿಸುವುದು
2. ಸಾಲಿನ ಸರಾಸರಿ ಮೌಲ್ಯದಿಂದ ಬದಲಾಯಿಸುವುದು

ನಾವು ಮಧ್ಯಮ ಮೌಲ್ಯದಿಂದ ಬದಲಾಯಿಸುವುದು, ಹೊರಗಿನ ಮೌಲ್ಯಗಳೊಂದಿಗೆ ತಿರುವುಳ್ಳ ಡೇಟಾದಲ್ಲಿ ಮಾಡುತ್ತೇವೆ. ಇದಕ್ಕೆ ಕಾರಣ ಮಧ್ಯಮವು ಹೊರಗಿನ ಮೌಲ್ಯಗಳಿಗೆ ಪ್ರತಿರೋಧಕವಾಗಿರುತ್ತದೆ.

ಡೇಟಾ ಸಾಮಾನ್ಯೀಕೃತವಾಗಿದ್ದಾಗ, ನಾವು ಸರಾಸರಿಯನ್ನು ಬಳಸಬಹುದು, ಏಕೆಂದರೆ ಆ ಸಂದರ್ಭದಲ್ಲಿ ಸರಾಸರಿ ಮತ್ತು ಮಧ್ಯಮವು ಬಹಳ ಸಮೀಪವಾಗಿರುತ್ತವೆ.

ಮೊದಲು, ನಾವು ಸಾಮಾನ್ಯವಾಗಿ ಹಂಚಿಕೆಯಾದ ಒಂದು ಕಾಲಮ್ ತೆಗೆದುಕೊಳ್ಳೋಣ ಮತ್ತು ಆ ಕಾಲಮ್‌ನ ಸರಾಸರಿ ಮೌಲ್ಯದಿಂದ ಕಾಣೆಯಾದ ಮೌಲ್ಯವನ್ನು ತುಂಬೋಣ.


In [32]:
fill_with_mean = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [np.nan,4,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  NaN  4  5
3  1.0  6  7
4  2.0  8  9

ಕಾಲಮ್‌ನ ಸರಾಸರಿ ಎಂದರೆ


In [33]:
np.mean(fill_with_mean[0])

0.0

ಸರಾಸರಿ ಬಳಸಿ ತುಂಬುವುದು


In [34]:
fill_with_mean[0].fillna(np.mean(fill_with_mean[0]),inplace=True)
fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  0.0  4  5
3  1.0  6  7
4  2.0  8  9

ನಾವು ನೋಡಬಹುದು, ಕಳೆದುಹೋಗಿರುವ ಮೌಲ್ಯವನ್ನು ಅದರ ಸರಾಸರಿಯಿಂದ ಬದಲಾಯಿಸಲಾಗಿದೆ.


ಈಗ ನಾವು ಮತ್ತೊಂದು ಡೇಟಾಫ್ರೇಮ್ ಅನ್ನು ಪ್ರಯತ್ನಿಸೋಣ, ಮತ್ತು ಈ ಬಾರಿ ನಾವು None ಮೌಲ್ಯಗಳನ್ನು ಕಾಲಮ್‌ನ ಮಧ್ಯಮ ಮೌಲ್ಯದಿಂದ ಬದಲಾಯಿಸುವೆವು.


In [35]:
fill_with_median = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [0,np.nan,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  NaN  5
3  1  6.0  7
4  2  8.0  9

ಎರಡನೇ ಕಾಲಮ್‌ನ ಮಧ್ಯಮ ಮೌಲ್ಯವೇನು


In [36]:
fill_with_median[1].median()

4.0

ಮಧ್ಯಮ ಮೌಲ್ಯದಿಂದ ತುಂಬುವುದು


In [37]:
fill_with_median[1].fillna(fill_with_median[1].median(),inplace=True)
fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  4.0  5
3  1  6.0  7
4  2  8.0  9

ನಾವು ನೋಡಬಹುದು, NaN ಮೌಲ್ಯವನ್ನು ಕಾಲಮ್‌ನ ಮಧ್ಯಮ ಮೌಲ್ಯದಿಂದ ಬದಲಾಯಿಸಲಾಗಿದೆ


In [38]:
example5 = pd.Series([1, np.nan, 2, None, 3], index=list('abcde'))
example5

a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64

ನೀವು ಎಲ್ಲಾ ಶೂನ್ಯ ಎಂಟ್ರಿಗಳನ್ನು ಒಂದೇ ಮೌಲ್ಯದಿಂದ ತುಂಬಬಹುದು, ಉದಾಹರಣೆಗೆ `0`:


In [39]:
example5.fillna(0)

a    1.0
b    0.0
c    2.0
d    0.0
e    3.0
dtype: float64

> ಮುಖ್ಯ ಅಂಶಗಳು:
1. ಕಡತದ ಮೌಲ್ಯಗಳನ್ನು ತುಂಬುವುದು ಕಡತದ ಡೇಟಾ ಕಡಿಮೆ ಇದ್ದಾಗ ಅಥವಾ ಕಡತದ ಡೇಟಾವನ್ನು ತುಂಬಲು ಯಾವುದೇ ತಂತ್ರವಿದ್ದಾಗ ಮಾಡಬೇಕು.
2. ಡೊಮೇನ್ ಜ್ಞಾನವನ್ನು ಬಳಸಿ ಕಡತದ ಮೌಲ್ಯಗಳನ್ನು ಅಂದಾಜು ಮಾಡಿ ತುಂಬಬಹುದು.
3. ವರ್ಗೀಕೃತ ಡೇಟಾದಲ್ಲಿ, ಬಹುಮಟ್ಟಿಗೆ, ಕಡತದ ಮೌಲ್ಯಗಳನ್ನು ಆ ಕಾಲಮ್‌ನ ಮೋಡ್‌ನಿಂದ ಬದಲಿಸಲಾಗುತ್ತದೆ.
4. ಸಂಖ್ಯಾತ್ಮಕ ಡೇಟಾದಲ್ಲಿ, ಕಡತದ ಮೌಲ್ಯಗಳನ್ನು ಸಾಮಾನ್ಯವಾಗಿ ಸರಾಸರಿ (ಸಾಮಾನ್ಯೀಕೃತ ಡೇಟಾಸೆಟ್‌ಗಳಿಗೆ) ಅಥವಾ ಕಾಲಮ್‌ಗಳ ಮಧ್ಯಮ ಮೌಲ್ಯದಿಂದ ತುಂಬಲಾಗುತ್ತದೆ.


### ವ್ಯಾಯಾಮ:


In [40]:
# What happens if you try to fill null values with a string, like ''?


ನೀವು **ಮುಂದಿನ-ಪೂರಣೆ** ಶೂನ್ಯ ಮೌಲ್ಯಗಳನ್ನು ಮಾಡಬಹುದು, ಅಂದರೆ ಶೂನ್ಯವನ್ನು ತುಂಬಲು ಕೊನೆಯ ಮಾನ್ಯ ಮೌಲ್ಯವನ್ನು ಬಳಸುವುದು:


In [41]:
example5.fillna(method='ffill')

a    1.0
b    1.0
c    2.0
d    2.0
e    3.0
dtype: float64

ನೀವು ಮುಂದಿನ ಮಾನ್ಯ ಮೌಲ್ಯವನ್ನು ಹಿಂಬಾಲಿಸಿ ಶೂನ್ಯವನ್ನು ತುಂಬಲು **ಬ್ಯಾಕ್-ಫಿಲ್** ಕೂಡ ಮಾಡಬಹುದು:


In [42]:
example5.fillna(method='bfill')

a    1.0
b    2.0
c    2.0
d    3.0
e    3.0
dtype: float64

ನೀವು ಊಹಿಸಬಹುದು, ಇದು DataFrames ಜೊತೆಗೆ ಕೂಡ ಇದೇ ರೀತಿಯಲ್ಲಿ ಕೆಲಸ ಮಾಡುತ್ತದೆ, ಆದರೆ ನೀವು null ಮೌಲ್ಯಗಳನ್ನು ತುಂಬಲು `axis` ಅನ್ನು ಕೂಡ ನಿರ್ದಿಷ್ಟಪಡಿಸಬಹುದು:


In [43]:
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

In [44]:
example4.fillna(method='ffill', axis=1)

0    1    2    3
0  1.0  1.0  7.0  7.0
1  2.0  5.0  8.0  8.0
2  NaN  6.0  9.0  9.0

ಹಿಂದಿನ ಮೌಲ್ಯವು ಮುಂದಿನ ಭರ್ತಿಗೆ ಲಭ್ಯವಿಲ್ಲದಿದ್ದಾಗ, ಶೂನ್ಯ ಮೌಲ್ಯ ಉಳಿಯುತ್ತದೆ ಎಂದು ಗಮನಿಸಿ.


### ವ್ಯಾಯಾಮ:


In [45]:
# What output does example4.fillna(method='bfill', axis=1) produce?
# What about example4.fillna(method='ffill') or example4.fillna(method='bfill')?
# Can you think of a longer code snippet to write that can fill all of the null values in example4?


ನೀವು `fillna` ಅನ್ನು ಬಳಸುವ ಬಗ್ಗೆ ಸೃಜನಶೀಲವಾಗಿರಬಹುದು. ಉದಾಹರಣೆಗೆ, ಮತ್ತೆ `example4` ಅನ್ನು ನೋಡೋಣ, ಆದರೆ ಈ ಬಾರಿ `DataFrame` ನಲ್ಲಿ ಇರುವ ಎಲ್ಲಾ ಮೌಲ್ಯಗಳ ಸರಾಸರಿಯನ್ನು ಬಳಸಿಕೊಂಡು ಕಳೆದುಹೋಗಿರುವ ಮೌಲ್ಯಗಳನ್ನು ತುಂಬೋಣ:


In [46]:
example4.fillna(example4.mean())

0    1  2   3
0  1.0  5.5  7 NaN
1  2.0  5.0  8 NaN
2  1.5  6.0  9 NaN

ಕಾಲಮ್ 3 ಇನ್ನೂ ಮೌಲ್ಯರಹಿತವಾಗಿದೆ ಎಂದು ಗಮನಿಸಿ: ಡೀಫಾಲ್ಟ್ ದಿಕ್ಕು ಮೌಲ್ಯಗಳನ್ನು ಸಾಲು-ವಾರಿಯಾಗಿ ತುಂಬುವುದು.

> **ಮುಖ್ಯಾಂಶ:** ನಿಮ್ಮ ಡೇಟಾಸೆಟ್‌ಗಳಲ್ಲಿ ಕಾಣಿಸದ ಮೌಲ್ಯಗಳನ್ನು ನಿರ್ವಹಿಸಲು ಹಲವಾರು ವಿಧಾನಗಳಿವೆ. ನೀವು ಬಳಸುವ ನಿರ್ದಿಷ್ಟ ತಂತ್ರ (ಅವುಗಳನ್ನು ತೆಗೆದುಹಾಕುವುದು, ಬದಲಾಯಿಸುವುದು, ಅಥವಾ ಬದಲಾಯಿಸುವ ವಿಧಾನವೇನಾಗಿರಲಿ) ಆ ಡೇಟಾದ ವಿಶೇಷತೆಯಿಂದ ನಿರ್ಧರಿಸಬೇಕು. ನೀವು ಡೇಟಾಸೆಟ್‌ಗಳನ್ನು ಹೆಚ್ಚು ಹ್ಯಾಂಡಲ್ ಮಾಡಿ, ಸಂವಹನ ಮಾಡುತ್ತಾ ಹೋಗುವಂತೆ ಕಾಣಿಸದ ಮೌಲ್ಯಗಳನ್ನು ನಿರ್ವಹಿಸುವ ಉತ್ತಮ ಜ್ಞಾನವನ್ನು ಅಭಿವೃದ್ಧಿಪಡಿಸುತ್ತೀರಿ.


### ವರ್ಗೀಕೃತ ಡೇಟಾವನ್ನು ಎನ್‌ಕೋಡ್ ಮಾಡುವುದು

ಯಂತ್ರ ಅಧ್ಯಯನ ಮಾದರಿಗಳು ಸಂಖ್ಯೆಗಳನ್ನೇ ಮಾತ್ರ ಹ್ಯಾಂಡಲ್ ಮಾಡುತ್ತವೆ ಮತ್ತು ಯಾವುದೇ ರೀತಿಯ ಸಂಖ್ಯಾತ್ಮಕ ಡೇಟಾವನ್ನೇ ಮಾತ್ರ. ಅದು ಹೌದು ಮತ್ತು ಇಲ್ಲದ ನಡುವಿನ ವ್ಯತ್ಯಾಸವನ್ನು ಹೇಳಲು ಸಾಧ್ಯವಿಲ್ಲ, ಆದರೆ 0 ಮತ್ತು 1 ನಡುವಿನ ವ್ಯತ್ಯಾಸವನ್ನು ಗುರುತಿಸಲು ಸಾಧ್ಯ. ಆದ್ದರಿಂದ, ಕಳೆದುಹೋಗಿರುವ ಮೌಲ್ಯಗಳನ್ನು ತುಂಬಿದ ನಂತರ, ಮಾದರಿ ಅರ್ಥಮಾಡಿಕೊಳ್ಳಲು ವರ್ಗೀಕೃತ ಡೇಟಾವನ್ನು ಕೆಲವು ಸಂಖ್ಯಾತ್ಮಕ ರೂಪಕ್ಕೆ ಎನ್‌ಕೋಡ್ ಮಾಡಬೇಕಾಗುತ್ತದೆ.

ಎನ್‌ಕೋಡಿಂಗ್ ಎರಡು ರೀತಿಗಳಲ್ಲಿ ಮಾಡಬಹುದು. ನಾವು ಅವುಗಳನ್ನು ಮುಂದಿನ ಭಾಗದಲ್ಲಿ ಚರ್ಚಿಸುವೆವು.


**ಲೇಬಲ್ ಎನ್‌ಕೋಡಿಂಗ್**


ಲೇಬಲ್ ಎನ್‌ಕೋಡಿಂಗ್ ಅಂದರೆ ಪ್ರತಿ ವರ್ಗವನ್ನು ಸಂಖ್ಯೆಗಾಗಿಸುವುದು. ಉದಾಹರಣೆಗೆ, ನಮ್ಮ ಬಳಿ ವಿಮಾನಯಾನ ಪ್ರಯಾಣಿಕರ ಡೇಟಾಸೆಟ್ ಇದ್ದು, ಅದರಲ್ಲೊಂದು ಕಾಲಮ್‌ನಲ್ಲಿ ಅವರ ವರ್ಗ ['ಬಿಸಿನೆಸ್ ಕ್ಲಾಸ್', 'ಇಕಾನಮಿ ಕ್ಲಾಸ್', 'ಫಸ್ಟ್ ಕ್ಲಾಸ್'] ಇರುತ್ತದೆ ಎಂದು ಹೇಳೋಣ. ಈ ಮೇಲೆ ಲೇಬಲ್ ಎನ್‌ಕೋಡಿಂಗ್ ಮಾಡಿದರೆ, ಇದು [0,1,2] ಆಗಿ ಪರಿವರ್ತಿತವಾಗುತ್ತದೆ. ಕೋಡ್ ಮೂಲಕ ಉದಾಹರಣೆಯನ್ನು ನೋಡೋಣ. ಮುಂದಿನ ನೋಟ್ಬುಕ್‌ಗಳಲ್ಲಿ ನಾವು `scikit-learn` ಅನ್ನು ಕಲಿಯಲಿದ್ದೇವೆ, ಆದ್ದರಿಂದ ಇಲ್ಲಿ ಅದನ್ನು ಬಳಸುವುದಿಲ್ಲ.


In [47]:
label = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
label

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

ಮೊದಲ ಕಾಲಮ್‌ನಲ್ಲಿ ಲೇಬಲ್ ಎನ್‌ಕೋಡಿಂಗ್ ಮಾಡಲು, ಪ್ರತಿಯೊಂದು ವರ್ಗವನ್ನು ಒಂದು ಸಂಖ್ಯೆಗೆ ನಕ್ಷೆ ಮಾಡಬೇಕಾಗುತ್ತದೆ, ಅದನ್ನು ಬದಲಾಯಿಸುವ ಮೊದಲು.


In [48]:
class_labels = {'business class':0,'economy class':1,'first class':2}
label['class'] = label['class'].replace(class_labels)
label

ID  class
0  10      0
1  20      2
2  30      1
3  40      1
4  50      1
5  60      0

ನಾವು ನೋಡಬಹುದು, ಔಟ್‌ಪುಟ್ ನಾವು ಭಾವಿಸಿದ್ದದಕ್ಕೆ ಹೊಂದಿಕೆಯಾಗುತ್ತದೆ. ಆದ್ದರಿಂದ, ನಾವು ಲೇಬಲ್ ಎನ್‌ಕೋಡಿಂಗ್ ಅನ್ನು ಯಾವಾಗ ಬಳಸುತ್ತೇವೆ? ಲೇಬಲ್ ಎನ್‌ಕೋಡಿಂಗ್ ಕೆಳಗಿನ ಯಾವುದೇ ಒಂದು ಅಥವಾ ಎರಡೂ ಸಂದರ್ಭಗಳಲ್ಲಿ ಬಳಸಲಾಗುತ್ತದೆ:
1. ವರ್ಗಗಳ ಸಂಖ್ಯೆ ಹೆಚ್ಚು ಇದ್ದಾಗ
2. ವರ್ಗಗಳು ಕ್ರಮದಲ್ಲಿ ಇದ್ದಾಗ.


**ಒನ್ ಹಾಟ್ ಎನ್‌ಕೋಡಿಂಗ್**

ಮತ್ತೊಂದು ಎನ್‌ಕೋಡಿಂಗ್ ಪ್ರಕಾರವು ಒನ್ ಹಾಟ್ ಎನ್‌ಕೋಡಿಂಗ್ ಆಗಿದೆ. ಈ ಎನ್‌ಕೋಡಿಂಗ್ ಪ್ರಕಾರದಲ್ಲಿ, ಕಾಲಮ್‌ನ ಪ್ರತಿ ವರ್ಗವನ್ನು ಪ್ರತ್ಯೇಕ ಕಾಲಮ್ ಆಗಿ ಸೇರಿಸಲಾಗುತ್ತದೆ ಮತ್ತು ಪ್ರತಿ ಡೇಟಾಪಾಯಿಂಟ್ ಆ ವರ್ಗವನ್ನು ಹೊಂದಿದೆಯೇ ಎಂಬುದರ ಆಧಾರದ ಮೇಲೆ 0 ಅಥವಾ 1 ಅನ್ನು ಪಡೆಯುತ್ತದೆ. ಆದ್ದರಿಂದ, n ವಿಭಿನ್ನ ವರ್ಗಗಳಿದ್ದರೆ, n ಕಾಲಮ್‌ಗಳು ಡೇಟಾಫ್ರೇಮ್‌ಗೆ ಸೇರಿಸಲಾಗುತ್ತದೆ.

ಉದಾಹರಣೆಗೆ, ನಾವು ಅದೇ ವಿಮಾನ ತರಗತಿ ಉದಾಹರಣೆಯನ್ನು ತೆಗೆದುಕೊಳ್ಳೋಣ. ವರ್ಗಗಳು ಇವು: ['ಬಿಸಿನೆಸ್ ಕ್ಲಾಸ್', 'ಇಕಾನಮಿ ಕ್ಲಾಸ್', 'ಫಸ್ಟ್ ಕ್ಲಾಸ್']. ಆದ್ದರಿಂದ, ಒನ್ ಹಾಟ್ ಎನ್‌ಕೋಡಿಂಗ್ ಮಾಡಿದರೆ, ಕೆಳಗಿನ ಮೂರು ಕಾಲಮ್‌ಗಳು ಡೇಟಾಸೆಟ್‌ಗೆ ಸೇರಿಸಲಾಗುತ್ತದೆ: ['class_ಬಿಸಿನೆಸ್ ಕ್ಲಾಸ್', 'class_ಇಕಾನಮಿ ಕ್ಲಾಸ್', 'class_ಫಸ್ಟ್ ಕ್ಲಾಸ್'].


In [49]:
one_hot = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
one_hot

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

ನಾವು ಮೊದಲ ಕಾಲಮ್ ಮೇಲೆ ಒನ್ ಹಾಟ್ ಎನ್‌ಕೋಡಿಂಗ್ ಮಾಡೋಣ


In [50]:
one_hot_data = pd.get_dummies(one_hot,columns=['class'])

In [51]:
one_hot_data

ID  class_business class  class_economy class  class_first class
0  10                     1                    0                  0
1  20                     0                    0                  1
2  30                     0                    1                  0
3  40                     0                    1                  0
4  50                     0                    1                  0
5  60                     1                    0                  0

ಪ್ರತಿ ಒನ್ ಹಾಟ್ ಎನ್‌ಕೋಡ್ ಮಾಡಿದ ಕಾಲಮ್ 0 ಅಥವಾ 1 ಅನ್ನು ಹೊಂದಿರುತ್ತದೆ, ಇದು ಆ ಡೇಟಾಪಾಯಿಂಟ್‌ಗೆ ಆ ವರ್ಗವು ಅಸ್ತಿತ್ವದಲ್ಲಿದೆಯೇ ಎಂದು ಸೂಚಿಸುತ್ತದೆ.


ನಾವು ಯಾವಾಗ ಒನ್ ಹಾಟ್ ಎನ್‌ಕೋಡಿಂಗ್ ಅನ್ನು ಬಳಸುತ್ತೇವೆ? ಒನ್ ಹಾಟ್ ಎನ್‌ಕೋಡಿಂಗ್ ಕೆಳಗಿನ ಯಾವುದೇ ಅಥವಾ ಎರಡೂ ಸಂದರ್ಭಗಳಲ್ಲಿ ಬಳಸಲಾಗುತ್ತದೆ:

1. ವರ್ಗಗಳ ಸಂಖ್ಯೆ ಮತ್ತು ಡೇಟಾಸೆಟ್‌ನ ಗಾತ್ರವು ಚಿಕ್ಕದಾಗಿರುವಾಗ.
2. ವರ್ಗಗಳು ಯಾವುದೇ ನಿರ್ದಿಷ್ಟ ಕ್ರಮವನ್ನು ಅನುಸರಿಸದಾಗ.


> ಮುಖ್ಯ ಅಂಶಗಳು:
1. ಎನ್ಕೋಡಿಂಗ್ ಅನ್ನು ಅಂಕಿ ರಹಿತ ಡೇಟಾವನ್ನು ಅಂಕಿ ಡೇಟಾಗೆ ಪರಿವರ್ತಿಸಲು ಮಾಡಲಾಗುತ್ತದೆ.
2. ಎನ್ಕೋಡಿಂಗ್‌ನ ಎರಡು ವಿಧಗಳಿವೆ: ಲೇಬಲ್ ಎನ್ಕೋಡಿಂಗ್ ಮತ್ತು ಒನ್ ಹಾಟ್ ಎನ್ಕೋಡಿಂಗ್, ಇವುಗಳನ್ನೆಲ್ಲಾ ಡೇಟಾಸೆಟ್‌ನ ಅಗತ್ಯಗಳ ಆಧಾರದ ಮೇಲೆ ಮಾಡಬಹುದು.


## ನಕಲಿ ಡೇಟಾವನ್ನು ತೆಗೆದುಹಾಕುವುದು

> **ಕಲಿಕೆಯ ಗುರಿ:** ಈ ಉಪವಿಭಾಗದ ಕೊನೆಯಲ್ಲಿ, ನೀವು ಡೇಟಾಫ್ರೇಮ್ಗಳಿಂದ ನಕಲಿ ಮೌಲ್ಯಗಳನ್ನು ಗುರುತಿಸುವುದು ಮತ್ತು ತೆಗೆದುಹಾಕುವುದರಲ್ಲಿ ಆರಾಮವಾಗಿರಬೇಕು.

ಕಳೆದುಹೋಗಿರುವ ಡೇಟಾದ ಜೊತೆಗೆ, ನೀವು ನಿಜಜೀವನದ ಡೇಟಾಸೆಟ್‌ಗಳಲ್ಲಿ ನಕಲಿ ಡೇಟಾವನ್ನು ಸಹ συχνά ಎದುರಿಸುತ್ತೀರಿ. ಭಾಗ್ಯವಶಾತ್, pandas ನಕಲಿ ಎಂಟ್ರಿಗಳನ್ನು ಪತ್ತೆಹಚ್ಚಲು ಮತ್ತು ತೆಗೆದುಹಾಕಲು ಸುಲಭವಾದ ವಿಧಾನವನ್ನು ಒದಗಿಸುತ್ತದೆ.


### ನಕಲುಗಳನ್ನು ಗುರುತಿಸುವುದು: `duplicated`

ನೀವು pandas ನಲ್ಲಿ `duplicated` ವಿಧಾನವನ್ನು ಬಳಸಿಕೊಂಡು ಸುಲಭವಾಗಿ ನಕಲು ಮೌಲ್ಯಗಳನ್ನು ಗುರುತಿಸಬಹುದು, ಇದು ಒಂದು ಬೂಲಿಯನ್ ಮಾಸ್ಕ್ ಅನ್ನು ಹಿಂತಿರುಗಿಸುತ್ತದೆ, ಅದು `DataFrame` ನಲ್ಲಿ ಒಂದು ಎಂಟ್ರಿ ಹಿಂದಿನದೊಂದಿಗಿನ ನಕಲು ಆಗಿದೆಯೇ ಎಂದು ಸೂಚಿಸುತ್ತದೆ. ಇದನ್ನು ಕಾರ್ಯಾಚರಣೆಯಲ್ಲಿ ನೋಡಲು ಮತ್ತೊಂದು ಉದಾಹರಣೆಯ `DataFrame` ಅನ್ನು ರಚಿಸೋಣ.


In [52]:
example6 = pd.DataFrame({'letters': ['A','B'] * 2 + ['B'],
                         'numbers': [1, 2, 1, 3, 3]})
example6

letters  numbers
0       A        1
1       B        2
2       A        1
3       B        3
4       B        3

In [53]:
example6.duplicated()

0    False
1    False
2     True
3    False
4     True
dtype: bool

### ನಕಲುಗಳನ್ನು ಬಿಟ್ಟಿಹಾಕುವುದು: `drop_duplicates`
`drop_duplicates` ಸರಳವಾಗಿ ಎಲ್ಲಾ `duplicated` ಮೌಲ್ಯಗಳು `False` ಆಗಿರುವ ಡೇಟಾದ ನಕಲನ್ನು ಹಿಂತಿರುಗಿಸುತ್ತದೆ:


In [54]:
example6.drop_duplicates()

letters  numbers
0       A        1
1       B        2
3       B        3

`duplicated` ಮತ್ತು `drop_duplicates` ಎರಡೂ ಡೀಫಾಲ್ಟ್ ಆಗಿ ಎಲ್ಲಾ ಕಾಲಮ್‌ಗಳನ್ನು ಪರಿಗಣಿಸುತ್ತವೆ ಆದರೆ ನೀವು ನಿಮ್ಮ `DataFrame` ನಲ್ಲಿ ಕೆಲವು ಕಾಲಮ್‌ಗಳ ಉಪಸಮೂಹವನ್ನು ಮಾತ್ರ ಪರಿಶೀಲಿಸಲು ಸೂಚಿಸಬಹುದು:


In [55]:
example6.drop_duplicates(['letters'])

letters  numbers
0       A        1
1       B        2

> **ಮುಖ್ಯಾಂಶ:** ನಕಲಿ ಡೇಟಾವನ್ನು ತೆಗೆದುಹಾಕುವುದು ಪ್ರಾಯೋಗಿಕವಾಗಿ ಪ್ರತಿಯೊಂದು ಡೇಟಾ-ವಿಜ್ಞಾನ ಯೋಜನೆಯಲ್ಲಿಯೂ ಅಗತ್ಯವಾದ ಭಾಗವಾಗಿದೆ. ನಕಲಿ ಡೇಟಾ ನಿಮ್ಮ ವಿಶ್ಲೇಷಣೆಯ ಫಲಿತಾಂಶಗಳನ್ನು ಬದಲಾಯಿಸಬಹುದು ಮತ್ತು ನಿಮಗೆ ತಪ್ಪು ಫಲಿತಾಂಶಗಳನ್ನು ನೀಡಬಹುದು!


## ನೈಜ-ಜಗತ್ತಿನ ಡೇಟಾ ಗುಣಮಟ್ಟ ಪರಿಶೀಲನೆಗಳು

> **ಕಲಿಕೆಯ ಗುರಿ:** ಈ ವಿಭಾಗದ ಕೊನೆಯಲ್ಲಿ, ನೀವು ಸಾಮಾನ್ಯ ನೈಜ-ಜಗತ್ತಿನ ಡೇಟಾ ಗುಣಮಟ್ಟ ಸಮಸ್ಯೆಗಳನ್ನು ಪತ್ತೆಹಚ್ಚಿ ಸರಿಪಡಿಸುವಲ್ಲಿ ಆರಾಮವಾಗಿರಬೇಕು, ಇದರಲ್ಲಿ ಅಸಮಾನ ವರ್ಗೀಕೃತ ಮೌಲ್ಯಗಳು, ಅಸಾಮಾನ್ಯ ಸಂಖ್ಯಾತ್ಮಕ ಮೌಲ್ಯಗಳು (ಔಟ್‌ಲೈಯರ್‌ಗಳು), ಮತ್ತು ಬದಲಾವಣೆಗಳೊಂದಿಗೆ ನಕಲಿ ಘಟಕಗಳು ಸೇರಿವೆ.

ಕಳೆದುಹೋಗಿರುವ ಮೌಲ್ಯಗಳು ಮತ್ತು ನಿಖರ ನಕಲುಗಳು ಸಾಮಾನ್ಯ ಸಮಸ್ಯೆಗಳಾಗಿದ್ದರೂ, ನೈಜ-ಜಗತ್ತಿನ ಡೇಟಾಸೆಟ್‌ಗಳು ಹೆಚ್ಚಾಗಿ ಸೂಕ್ಷ್ಮ ಸಮಸ್ಯೆಗಳನ್ನು ಹೊಂದಿರುತ್ತವೆ:

1. **ಅಸಮಾನ ವರ್ಗೀಕೃತ ಮೌಲ್ಯಗಳು**: ಒಂದೇ ವರ್ಗವನ್ನು ವಿಭಿನ್ನವಾಗಿ ಬರೆಯುವುದು (ಉದಾ: "USA", "U.S.A", "United States")
2. **ಅಸಾಮಾನ್ಯ ಸಂಖ್ಯಾತ್ಮಕ ಮೌಲ್ಯಗಳು**: ಡೇಟಾ ನಮೂದಿಸುವ ದೋಷಗಳನ್ನು ಸೂಚಿಸುವ ಅತಿದೊಡ್ಡ ಔಟ್‌ಲೈಯರ್‌ಗಳು (ಉದಾ: ವಯಸ್ಸು = 999)
3. **ಸಮೀಪ ನಕಲಿ ಸಾಲುಗಳು**: ಸ್ವಲ್ಪ ಬದಲಾವಣೆಗಳೊಂದಿಗೆ ಒಂದೇ ಘಟಕವನ್ನು ಪ್ರತಿನಿಧಿಸುವ ದಾಖಲೆಗಳು

ಈ ಸಮಸ್ಯೆಗಳನ್ನು ಪತ್ತೆಹಚ್ಚಿ ನಿರ್ವಹಿಸುವ ತಂತ್ರಗಳನ್ನು ಅನ್ವೇಷಿಸೋಣ.


### ಮಾದರಿ "ಕಳಪೆ" ಡೇಟಾಸೆಟ್ ರಚನೆ

ಮೊದಲು, ನಾವು ಸಾಮಾನ್ಯವಾಗಿ ನಿಜವಾದ ಡೇಟಾದಲ್ಲಿ ಎದುರಿಸುವ ಸಮಸ್ಯೆಗಳ ಪ್ರಕಾರಗಳನ್ನೊಳಗೊಂಡ ಮಾದರಿ ಡೇಟಾಸೆಟ್ ಅನ್ನು ರಚಿಸೋಣ:


In [ ]:
import pandas as pd
import numpy as np

# Create a sample dataset with quality issues
dirty_data = pd.DataFrame({
    'customer_id': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
    'name': ['John Smith', 'Jane Doe', 'John Smith', 'Bob Johnson', 
             'Alice Williams', 'Charlie Brown', 'John  Smith', 'Eva Martinez',
             'Bob Johnson', 'Diana Prince', 'Frank Castle', 'Alice Williams'],
    'age': [25, 32, 25, 45, 28, 199, 25, 31, 45, 27, -5, 28],
    'country': ['USA', 'UK', 'U.S.A', 'Canada', 'USA', 'United Kingdom',
                'United States', 'Mexico', 'canada', 'USA', 'UK', 'usa'],
    'purchase_amount': [100.50, 250.00, 105.00, 320.00, 180.00, 90.00,
                       102.00, 275.00, 325.00, 195.00, 410.00, 185.00]
})

print("Sample 'Dirty' Dataset:")
print(dirty_data)

### 1. ಅಸಮಾನ ವರ್ಗೀಕೃತ ಮೌಲ್ಯಗಳನ್ನು ಪತ್ತೆಹಚ್ಚುವುದು

`country` ಕಾಲಮ್‌ನಲ್ಲಿ ಒಂದೇ ದೇಶಗಳಿಗೆ ಹಲವಾರು ಪ್ರತಿನಿಧಾನಗಳಿವೆ ಎಂದು ಗಮನಿಸಿ. ಈ ಅಸಮಾನತೆಗಳನ್ನು ಗುರುತಿಸೋಣ:


In [ ]:
# Check unique values in the country column
print("Unique country values:")
print(dirty_data['country'].unique())
print(f"\nTotal unique values: {dirty_data['country'].nunique()}")

# Count occurrences of each variation
print("\nValue counts:")
print(dirty_data['country'].value_counts())

#### ವರ್ಗೀಕೃತ ಮೌಲ್ಯಗಳನ್ನು ಮಾನಕೀಕರಿಸುವುದು

ನಾವು ಈ ಮೌಲ್ಯಗಳನ್ನು ಮಾನಕೀಕರಿಸಲು ಒಂದು ನಕ್ಷೆ ರಚಿಸಬಹುದು. ಸರಳ ವಿಧಾನವೆಂದರೆ ಸಣ್ಣ ಅಕ್ಷರಗಳಿಗೆ ಪರಿವರ್ತಿಸಿ ಮತ್ತು ನಕ್ಷೆ ಡಿಕ್ಷನರಿ ರಚಿಸುವುದು:


In [ ]:
# Create a standardization mapping
country_mapping = {
    'usa': 'USA',
    'u.s.a': 'USA',
    'united states': 'USA',
    'uk': 'UK',
    'united kingdom': 'UK',
    'canada': 'Canada',
    'mexico': 'Mexico'
}

# Standardize the country column
dirty_data['country_clean'] = dirty_data['country'].str.lower().map(country_mapping)

print("Before standardization:")
print(dirty_data['country'].value_counts())
print("\nAfter standardization:")
print(dirty_data[['country_clean']].value_counts())

**ವೈಕಲ್ಪಿಕ: ಫಜ್ಜಿ ಹೊಂದಾಣಿಕೆ ಬಳಕೆ**

ಹೆಚ್ಚು ಸಂಕೀರ್ಣ ಪ್ರಕರಣಗಳಿಗೆ, ನಾವು `rapidfuzz` ಗ್ರಂಥಾಲಯದೊಂದಿಗೆ ಫಜ್ಜಿ ಸ್ಟ್ರಿಂಗ್ ಹೊಂದಾಣಿಕೆಯನ್ನು ಬಳಸಿಕೊಂಡು ಸ್ವಯಂಚಾಲಿತವಾಗಿ ಸಮಾನ ಸ್ಟ್ರಿಂಗ್‌ಗಳನ್ನು ಪತ್ತೆಹಚ್ಚಬಹುದು:


In [ ]:
try:
    from rapidfuzz import process, fuzz
except ImportError:
    print("rapidfuzz is not installed. Please install it with 'pip install rapidfuzz' to use fuzzy matching.")
    process = None
    fuzz = None

# Get unique countries
unique_countries = dirty_data['country'].unique()

# For each country, find similar matches
if process is not None and fuzz is not None:
    print("Finding similar country names (similarity > 70%):")
    for country in unique_countries:
        matches = process.extract(country, unique_countries, scorer=fuzz.ratio, limit=3)
        # Filter matches with similarity > 70 and not identical
        similar = [m for m in matches if m[1] > 70 and m[0] != country]
        if similar:
            print(f"\n'{country}' is similar to:")
            for match, score, _ in similar:
                print(f"  - '{match}' (similarity: {score}%)")
else:
    print("Skipping fuzzy matching because rapidfuzz is not available.")

### 2. ಅಸಾಮಾನ್ಯ ಸಂಖ್ಯಾತ್ಮಕ ಮೌಲ್ಯಗಳನ್ನು ಪತ್ತೆಹಚ್ಚುವುದು (ಔಟ್‌ಲೈಯರ್‌ಗಳು)

`age` ಕಾಲಮ್ ಅನ್ನು ನೋಡಿದಾಗ, 199 ಮತ್ತು -5 ಎಂಬ ಕೆಲವು ಸಂಶಯಾಸ್ಪದ ಮೌಲ್ಯಗಳಿವೆ. ಈ ಔಟ್‌ಲೈಯರ್‌ಗಳನ್ನು ಪತ್ತೆಹಚ್ಚಲು ನಾವು ಸಂಖ್ಯಾಶಾಸ್ತ್ರೀಯ ವಿಧಾನಗಳನ್ನು ಬಳಸೋಣ.


In [ ]:
# Display basic statistics
print("Age column statistics:")
print(dirty_data['age'].describe())

# Identify impossible values using domain knowledge
print("\nRows with impossible age values (< 0 or > 120):")
impossible_ages = dirty_data[(dirty_data['age'] < 0) | (dirty_data['age'] > 120)]
print(impossible_ages[['customer_id', 'name', 'age']])

#### IQR (ಇಂಟರ್‌ಕ್ವಾರ್ಟೈಲ್ ರೇಂಜ್) ವಿಧಾನವನ್ನು ಬಳಸುವುದು

IQR ವಿಧಾನವು ಅತಿದೊಡ್ಡ ಮೌಲ್ಯಗಳಿಗೆ ಕಡಿಮೆ ಸಂವೇದನಶೀಲವಾಗಿರುವ ಬಲವಾದ ಸಂಖ್ಯಾಶಾಸ್ತ್ರೀಯ ತಂತ್ರವಾಗಿದೆ:


In [ ]:
# Calculate IQR for age (excluding impossible values)
valid_ages = dirty_data[(dirty_data['age'] >= 0) & (dirty_data['age'] <= 120)]['age']

Q1 = valid_ages.quantile(0.25)
Q3 = valid_ages.quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print(f"IQR-based outlier bounds for age: [{lower_bound:.2f}, {upper_bound:.2f}]")

# Identify outliers
age_outliers = dirty_data[(dirty_data['age'] < lower_bound) | (dirty_data['age'] > upper_bound)]
print(f"\nRows with age outliers:")
print(age_outliers[['customer_id', 'name', 'age']])

#### Z-ಸ್ಕೋರ್ ವಿಧಾನವನ್ನು ಬಳಸುವುದು

Z-ಸ್ಕೋರ್ ವಿಧಾನವು ಸರಾಸರಿ ಇಂದ ಮಾನಕ ವ್ಯತ್ಯಾಸಗಳ ಆಧಾರದ ಮೇಲೆ ಹೊರಗಿನ ಅಂಕಿಗಳನ್ನು ಗುರುತಿಸುತ್ತದೆ:


In [ ]:
try:
    from scipy import stats
except ImportError:
    print("scipy is required for Z-score calculation. Please install it with 'pip install scipy' and rerun this cell.")
else:
    # Calculate Z-scores for age, handling NaN values
    age_nonan = dirty_data['age'].dropna()
    zscores = np.abs(stats.zscore(age_nonan))
    dirty_data['age_zscore'] = np.nan
    dirty_data.loc[age_nonan.index, 'age_zscore'] = zscores

    # Typically, Z-score > 3 indicates an outlier
    print("Rows with age Z-score > 3:")
    zscore_outliers = dirty_data[dirty_data['age_zscore'] > 3]
    print(zscore_outliers[['customer_id', 'name', 'age', 'age_zscore']])

    # Clean up the temporary column
    dirty_data = dirty_data.drop('age_zscore', axis=1)

#### ಹೊರಗಿನ ಅಂಶಗಳನ್ನು ನಿರ್ವಹಿಸುವುದು

ಒಮ್ಮೆ ಪತ್ತೆಯಾದ ನಂತರ, ಹೊರಗಿನ ಅಂಶಗಳನ್ನು ಹಲವಾರು ರೀತಿಯಲ್ಲಿ ನಿರ್ವಹಿಸಬಹುದು:
1. **ತೆಗೆದುಹಾಕಿ**: ಹೊರಗಿನ ಅಂಶಗಳಿರುವ ಸಾಲುಗಳನ್ನು ತೆಗೆದುಹಾಕಿ (ಅವು ದೋಷಗಳಾಗಿದ್ದರೆ)
2. **ಮಿತಿ ನಿಗದಿ ಮಾಡಿ**: ಗಡಿಬಿಡಿ ಮೌಲ್ಯಗಳಿಂದ ಬದಲಾಯಿಸಿ
3. **NaN ನೊಂದಿಗೆ ಬದಲಾಯಿಸಿ**: ಇಲ್ಲದಿರುವ ಡೇಟಾ ಎಂದು ಪರಿಗಣಿಸಿ ಮತ್ತು ಇಂಪುಟೇಶನ್ ತಂತ್ರಗಳನ್ನು ಬಳಸಿ
4. **ಉಳಿಸಿ**: ಅವು ಮಾನ್ಯವಾದ ಅತಿದೊಡ್ಡ ಮೌಲ್ಯಗಳಾಗಿದ್ದರೆ


In [ ]:
# Create a cleaned version by replacing impossible ages with NaN
dirty_data['age_clean'] = dirty_data['age'].apply(
    lambda x: np.nan if (x < 0 or x > 120) else x
)

print("Age column before and after cleaning:")
print(dirty_data[['customer_id', 'name', 'age', 'age_clean']])

### 3. ಸಮೀಪದ ನಕಲಿ ಸಾಲುಗಳನ್ನು ಪತ್ತೆಹಚ್ಚುವುದು

ನಮ್ಮ ಡೇಟಾಸೆಟ್‌ನಲ್ಲಿ "ಜಾನ್ ಸ್ಮಿತ್" ಎಂಬ ಹೆಸರಿನ ವಿವಿಧ ಮೌಲ್ಯಗಳೊಂದಿಗೆ ಹಲವಾರು ದಾಖಲೆಗಳಿವೆ ಎಂದು ಗಮನಿಸಿ. ಹೆಸರಿನ ಸಾದೃಶ್ಯದ ಆಧಾರದ ಮೇಲೆ ಸಾಧ್ಯವಿರುವ ನಕಲುಗಳನ್ನು ಗುರುತಿಸೋಣ.


In [ ]:
# First, let's look at exact name matches (ignoring extra whitespace)
dirty_data['name_normalized'] = dirty_data['name'].str.strip().str.lower()

print("Checking for duplicate names:")
duplicate_names = dirty_data[dirty_data.duplicated(['name_normalized'], keep=False)]
print(duplicate_names.sort_values('name_normalized')[['customer_id', 'name', 'age', 'country']])

#### ಫಜ್ಜಿ ಹೊಂದಾಣಿಕೆಯಿಂದ ಸಮೀಪದ ನಕಲುಗಳನ್ನು ಹುಡುಕುವುದು

ಹೆಚ್ಚು ಸುಧಾರಿತ ನಕಲು ಪತ್ತೆಗಾಗಿ, ನಾವು ಸಮಾನ ಹೆಸರುಗಳನ್ನು ಹುಡುಕಲು ಫಜ್ಜಿ ಹೊಂದಾಣಿಕೆಯನ್ನು ಬಳಸಬಹುದು:


In [ ]:
try:
    from rapidfuzz import process, fuzz

    # Function to find potential duplicates
    def find_near_duplicates(df, column, threshold=90):
        """
        Find near-duplicate entries in a column using fuzzy matching.
        
        Parameters:
        - df: DataFrame
        - column: Column name to check for duplicates
        - threshold: Similarity threshold (0-100)
        
        Returns: List of potential duplicate groups
        """
        values = df[column].unique()
        duplicate_groups = []
        checked = set()
        
        for value in values:
            if value in checked:
                continue
                
            # Find similar values
            matches = process.extract(value, values, scorer=fuzz.ratio, limit=len(values))
            similar = [m[0] for m in matches if m[1] >= threshold]
            
            if len(similar) > 1:
                duplicate_groups.append(similar)
                checked.update(similar)
        
        return duplicate_groups

    # Find near-duplicate names
    duplicate_groups = find_near_duplicates(dirty_data, 'name', threshold=90)

    print("Potential duplicate groups:")
    for i, group in enumerate(duplicate_groups, 1):
        print(f"\nGroup {i}:")
        for name in group:
            matching_rows = dirty_data[dirty_data['name'] == name]
            print(f"  '{name}': {len(matching_rows)} occurrence(s)")
            for _, row in matching_rows.iterrows():
                print(f"    - Customer {row['customer_id']}: age={row['age']}, country={row['country']}")
except ImportError:
    print("rapidfuzz is not installed. Skipping fuzzy matching for near-duplicates.")

#### ನಕಲುಗಳನ್ನು ನಿರ್ವಹಿಸುವುದು

ಒಮ್ಮೆ ಗುರುತಿಸಿದ ನಂತರ, ನಕಲುಗಳನ್ನು ಹೇಗೆ ನಿರ್ವಹಿಸುವುದೆಂದು ನೀವು ನಿರ್ಧರಿಸಬೇಕು:
1. **ಮೊದಲ ಸಂಭವನೆಯನ್ನು ಉಳಿಸಿ**: `drop_duplicates(keep='first')` ಬಳಸಿ
2. **ಕೊನೆಯ ಸಂಭವನೆಯನ್ನು ಉಳಿಸಿ**: `drop_duplicates(keep='last')` ಬಳಸಿ
3. **ಮಾಹಿತಿಯನ್ನು ಸಂಗ್ರಹಿಸಿ**: ನಕಲು ಸಾಲುಗಳಿಂದ ಮಾಹಿತಿಯನ್ನು ಸಂಯೋಜಿಸಿ
4. **ಮ್ಯಾನುಯಲ್ ಪರಿಶೀಲನೆ**: ಮಾನವ ಪರಿಶೀಲನೆಗಾಗಿ ಫ್ಲ್ಯಾಗ್ ಮಾಡಿ


In [ ]:
# Example: Remove duplicates based on normalized name, keeping first occurrence
cleaned_data = dirty_data.drop_duplicates(subset=['name_normalized'], keep='first')

print(f"Original dataset: {len(dirty_data)} rows")
print(f"After removing name duplicates: {len(cleaned_data)} rows")
print(f"Removed: {len(dirty_data) - len(cleaned_data)} duplicate rows")

print("\nCleaned dataset:")
print(cleaned_data[['customer_id', 'name', 'age', 'country_clean']])

### ಸಾರಾಂಶ: ಸಂಪೂರ್ಣ ಡೇಟಾ ಶುದ್ಧೀಕರಣ ಪೈಪ್‌ಲೈನ್

ನಾವು ಇದನ್ನು ಎಲ್ಲವನ್ನೂ ಸೇರಿಸಿ ಸಮಗ್ರ ಶುದ್ಧೀಕರಣ ಪೈಪ್‌ಲೈನ್ ಆಗಿ ಮಾಡೋಣ:


In [ ]:
def clean_dataset(df):
    """
    Comprehensive data cleaning function.
    """
    # Create a copy to avoid modifying the original
    cleaned = df.copy()
    
    # 1. Standardize categorical values (country)
    country_mapping = {
        'usa': 'USA', 'u.s.a': 'USA', 'united states': 'USA',
        'uk': 'UK', 'united kingdom': 'UK',
        'canada': 'Canada', 'mexico': 'Mexico'
    }
    cleaned['country'] = cleaned['country'].str.lower().map(country_mapping)
    
    # 2. Clean abnormal age values
    cleaned['age'] = cleaned['age'].apply(
        lambda x: np.nan if (x < 0 or x > 120) else x
    )
    
    # 3. Remove near-duplicate names (normalize whitespace)
    cleaned['name'] = cleaned['name'].str.strip()
    cleaned = cleaned.drop_duplicates(subset=['name'], keep='first')
    
    return cleaned

# Apply the cleaning pipeline
final_cleaned_data = clean_dataset(dirty_data)

print("Before cleaning:")
print(f"  Rows: {len(dirty_data)}")
print(f"  Unique countries: {dirty_data['country'].nunique()}")
print(f"  Invalid ages: {((dirty_data['age'] < 0) | (dirty_data['age'] > 120)).sum()}")

print("\nAfter cleaning:")
print(f"  Rows: {len(final_cleaned_data)}")
print(f"  Unique countries: {final_cleaned_data['country'].nunique()}")
print(f"  Invalid ages: {((final_cleaned_data['age'] < 0) | (final_cleaned_data['age'] > 120)).sum()}")

print("\nCleaned dataset:")
print(final_cleaned_data[['customer_id', 'name', 'age', 'country', 'purchase_amount']])

### 🎯 ಚಾಲೆಂಜ್ ವ್ಯಾಯಾಮ

ಈಗ ನಿಮ್ಮ ತಿರುಗು! ಕೆಳಗಿನವು ಗುಣಮಟ್ಟದ ಹಲವು ಸಮಸ್ಯೆಗಳೊಂದಿಗೆ ಹೊಸ ಸಾಲು ಡೇಟಾ. ನೀವು ಮಾಡಬಹುದೇ:

1. ಈ ಸಾಲಿನಲ್ಲಿ ಎಲ್ಲಾ ಸಮಸ್ಯೆಗಳನ್ನು ಗುರುತಿಸಿ
2. ಪ್ರತಿ ಸಮಸ್ಯೆಯನ್ನು ಶುದ್ಧಗೊಳಿಸಲು ಕೋಡ್ ಬರೆಯಿರಿ
3. ಶುದ್ಧಗೊಳಿಸಿದ ಸಾಲನ್ನು ಡೇಟಾಸೆಟ್‌ಗೆ ಸೇರಿಸಿ

ಇದು ಸಮಸ್ಯೆಯ ಡೇಟಾ:


In [ ]:
# New problematic row
new_row = pd.DataFrame({
    'customer_id': [13],
    'name': ['  Diana  Prince  '],  # Extra whitespace
    'age': [250],  # Impossible age
    'country': ['U.S.A.'],  # Inconsistent format
    'purchase_amount': [150.00]
})

print("New row to clean:")
print(new_row)

# TODO: Your code here to clean this row
# Hints:
# 1. Strip whitespace from the name
# 2. Check if the name is a duplicate (Diana Prince already exists)
# 3. Handle the impossible age value
# 4. Standardize the country name

# Example solution (uncomment and modify as needed):
# new_row_cleaned = new_row.copy()
# new_row_cleaned['name'] = new_row_cleaned['name'].str.strip()
# new_row_cleaned['age'] = np.nan  # Invalid age
# new_row_cleaned['country'] = 'USA'  # Standardized
# print("\nCleaned row:")
# print(new_row_cleaned)

### ಪ್ರಮುಖ ಅಂಶಗಳು

1. **ಅಸಮಾನ ವರ್ಗಗಳು** ನೈಜ ಜಗತ್ತಿನ ಡೇಟಾದಲ್ಲಿ ಸಾಮಾನ್ಯವಾಗಿವೆ. ಸದಾ ವಿಶಿಷ್ಟ ಮೌಲ್ಯಗಳನ್ನು ಪರಿಶೀಲಿಸಿ ಮತ್ತು ನಕ್ಷೆಗಳನ್ನು ಅಥವಾ ಫಜ್ಜಿ ಹೊಂದಾಣಿಕೆಯನ್ನು ಬಳಸಿ ಅವುಗಳನ್ನು ಮಾನಕೀಕರಿಸಿ.

2. **ಅತಿರೇಕಗಳು** ನಿಮ್ಮ ವಿಶ್ಲೇಷಣೆಯನ್ನು ಬಹುಮಟ್ಟಿಗೆ ಪ್ರಭಾವಿತ ಮಾಡಬಹುದು. ಅವುಗಳನ್ನು ಪತ್ತೆಹಚ್ಚಲು ಕ್ಷೇತ್ರ ಜ್ಞಾನವನ್ನು ಸಂಖ್ಯಾಶಾಸ್ತ್ರೀಯ ವಿಧಾನಗಳೊಂದಿಗೆ (IQR, Z-ಸ್ಕೋರ್) ಬಳಸಿ.

3. **ಸಮೀಪದ ನಕಲುಗಳು** ನಿಖರ ನಕಲುಗಳಿಗಿಂತ ಪತ್ತೆಹಚ್ಚಲು ಕಷ್ಟಕರವಾಗಿವೆ. ಅವುಗಳನ್ನು ಗುರುತಿಸಲು ಫಜ್ಜಿ ಹೊಂದಾಣಿಕೆ ಮತ್ತು ಡೇಟಾವನ್ನು ಸಾಮಾನ್ಯೀಕರಿಸುವುದು (ಕೆಳಗಿನ ಅಕ್ಷರಗೊಳಿಸುವುದು, ಖಾಲಿ ಸ್ಥಳಗಳನ್ನು ತೆಗೆದುಹಾಕುವುದು) ಪರಿಗಣಿಸಿ.

4. **ಡೇಟಾ ಶುದ್ಧೀಕರಣವು ಪುನರಾವರ್ತನಾತ್ಮಕವಾಗಿದೆ**. ನೀವು ಅನೇಕ ತಂತ್ರಗಳನ್ನು ಅನ್ವಯಿಸಬೇಕಾಗಬಹುದು ಮತ್ತು ನಿಮ್ಮ ಶುದ್ಧೀಕೃತ ಡೇಟಾಸೆಟ್ ಅನ್ನು ಅಂತಿಮಗೊಳಿಸುವ ಮೊದಲು ಫಲಿತಾಂಶಗಳನ್ನು ಪರಿಶೀಲಿಸಬೇಕಾಗಬಹುದು.

5. **ನಿಮ್ಮ ನಿರ್ಧಾರಗಳನ್ನು ದಾಖಲೆ ಮಾಡಿಕೊಳ್ಳಿ**. ನೀವು ಯಾವ ಶುದ್ಧೀಕರಣ ಹಂತಗಳನ್ನು ಅನ್ವಯಿಸಿದ್ದೀರಿ ಮತ್ತು ಏಕೆ ಎಂದು ಗಮನದಲ್ಲಿಡಿ, ಏಕೆಂದರೆ ಇದು ಪುನರಾವರ್ತನೆ ಮತ್ತು ಪಾರದರ್ಶಕತೆಯಿಗಾಗಿ ಮುಖ್ಯವಾಗಿದೆ.

> **ಉತ್ತಮ ಅಭ್ಯಾಸ:** ನಿಮ್ಮ ಮೂಲ "ಕಳಪೆ" ಡೇಟಾದ ಪ್ರತಿಯನ್ನು ಸದಾ ಇಟ್ಟುಕೊಳ್ಳಿ. ನಿಮ್ಮ ಮೂಲ ಡೇಟಾ ಫೈಲ್‌ಗಳನ್ನು ಮರುಬರೆಯಬೇಡಿ - `data_cleaned.csv` ಹೋಲುವ ಸ್ಪಷ್ಟ ಹೆಸರಿನ ನಿಯಮಗಳೊಂದಿಗೆ ಶುದ್ಧೀಕೃತ ಆವೃತ್ತಿಗಳನ್ನು ರಚಿಸಿ.


---

<!-- CO-OP TRANSLATOR DISCLAIMER START -->
**ಅಸ್ವೀಕರಣ**:  
ಈ ದಸ್ತಾವೇಜು AI ಅನುವಾದ ಸೇವೆ [Co-op Translator](https://github.com/Azure/co-op-translator) ಬಳಸಿ ಅನುವಾದಿಸಲಾಗಿದೆ. ನಾವು ನಿಖರತೆಯಿಗಾಗಿ ಪ್ರಯತ್ನಿಸುತ್ತಿದ್ದರೂ, ಸ್ವಯಂಚಾಲಿತ ಅನುವಾದಗಳಲ್ಲಿ ದೋಷಗಳು ಅಥವಾ ಅಸತ್ಯತೆಗಳು ಇರಬಹುದು ಎಂದು ದಯವಿಟ್ಟು ಗಮನಿಸಿ. ಮೂಲ ಭಾಷೆಯಲ್ಲಿರುವ ಮೂಲ ದಸ್ತಾವೇಜನ್ನು ಅಧಿಕೃತ ಮೂಲವೆಂದು ಪರಿಗಣಿಸಬೇಕು. ಮಹತ್ವದ ಮಾಹಿತಿಗಾಗಿ, ವೃತ್ತಿಪರ ಮಾನವ ಅನುವಾದವನ್ನು ಶಿಫಾರಸು ಮಾಡಲಾಗುತ್ತದೆ. ಈ ಅನುವಾದ ಬಳಕೆಯಿಂದ ಉಂಟಾಗುವ ಯಾವುದೇ ತಪ್ಪು ಅರ್ಥಮಾಡಿಕೊಳ್ಳುವಿಕೆ ಅಥವಾ ತಪ್ಪು ವಿವರಣೆಗಳಿಗೆ ನಾವು ಹೊಣೆಗಾರರಾಗುವುದಿಲ್ಲ.
<!-- CO-OP TRANSLATOR DISCLAIMER END -->
